---
title: Large-scale chromatin accessibility profiling uncovers heterogeneity of molecular phenotypes and gene regulatory networks of CLL
author:
        - André F. Rendeiro^1,\*^,
        - Christian Schmidl^1,\*^,
        - Renata Walewska^2^,
        - Zadie Davis^2^,
        - Matthias Farlik^1^,
        - Jonathan Strefford^3^,
        - David Oscier^2^,
        - Christoph Bock^1,4,5^
        - ^1^CeMM Research Center for Molecular Medicine of the Austrian Academy of Sciences, Vienna, Austria
        - ^2^Department of Molecular Pathology, Royal Bournemouth Hospital, Bournemouth, UK
        - ^3^Cancer Genomics, Cancer Sciences, University of Southampton, Southampton, UK
        - ^4^Department of Laboratory Medicine, Medical University of Vienna, Vienna, Austria
        - ^5^Max Planck Institute for Informatics, Saarbrücken, Germany
        - ^\*^equal contribution
tags: [chronic lymphocytic leukaemia, chromatin, ATAC-seq, gene regulatory networks]
bibliography: /home/afr/Documents/library.bib
---

#### Key points:
- First chromatin accessibility profiling in a large cancer cohort identifies epigenetic heterogeneity of relevance to the disease
- Differential usage of open-chromatin regions between molecular subgroups reveals molecular signatures with alternative signaling pathway usage
- Gene regulatory network inference unveils alternate regulatory interactions between disease subtypes

## Abstract:
Recent advances in fast, low-input and cost-effective epigenome profiling technologies allow more comprehensive enquiry into the diversity of chromatin regulation in large cohorts of patients. By performing ATAC-seq on a large cohort of Chronic Lymphocytic Leukemia (CLL) cases, we characterized the chromatin landscape of CLL, finding substantial heterogeneity. With a supervised machine learning approach guided on clinical annotations, we uncovered regulatory regions associated with the IGHV mutation status – the largest molecular determinant of CLL classification. Differential usage of these regulatory regions reflects the likely cell-of-origin of each CLL group as these regions are associated with enhancers and promoters important for B cell differentiation. Moreover, enrichment analysis of regions-associated genes reveals differential pathways that may be driving disease progression and contributing to the distinction between aggressive and indolent cases. This classification approach allowed us to find additional molecular subgroups of CLL that we used to stratify patients, and which highlight the heterogeneity of CLL at the chromatin level. To gain deeper insights into the molecular wiring of CLL, we inferred gene regulatory networks (GRN) from transcription factor footprints, and describe the first GRN of this kind for a cancer. Comparing the inferred interaction networks from the previously identified molecular subgroups, we uncover novel differentially used network modules of potential therapeutic relevance.

## Introduction
Chronic lymphocytic leukemia (CLL) is one of the most common leukemia in adults, causing exaggerated proliferation of B lymphocytes in the bone marrow, lymph nodes and blood of patients. Two molecular subgroups of the disease exist, which are associated with the degree of somatic mutational in the immunoglobulin heavy chain variable region (IGHV) locus – a genetic scar of class switch recombination (CSR) in B cells during maturation. Cases with increased mutation of IGHV (mCLL) are more likely to have a mostly indolent form of the disease, whereas patients with a lowly mutated or canonical IGHV locus have worse prognosis [@Zenz2010]. This difference in mutational status of the IGHV locus in CLL is likely to reflect the cellular origin of the leukemic cells, due to the resemblance with naïve B cells bearing unmutated IGHV, and memory B cells, which have undergone CSR and therefore have a IGHV locus with low homology to the canonical version, although the exact origin of the normal B cell giving rise to CLL is still under debate [@Seifert2012].
Despite considerable advances in the understanding of the molecular pathology and in treatment, CLL remains incurable. Investigations on the genetic basis of CLL with large-scale exome- and whole genome sequencing [@Puente2011; @Quesada2011; @Puente2015; @Landau2015] has revealed a set of 44 recurrently mutated genes 6, but nonetheless, only 3 genes (TP53, SF3B1 and RPS15)  have showed significant differences in overall or progression-free survival of CLL patients [@Landau2015; @Baliakas2014].

Epigenetic marks can be seen as te manifestation of a cellular identity and state, restricting the differentiation potential of a cell. In light of thus, it is not surprising that epigenetic alterations are ubiquitous in cancer [@Yang2015], a disease which can be seen as a deregulation of cellular identity. Previous studies of epigenetics in the context of CLL, have focused on DNA methylation [@Kulis2012; @Pei2012; @Oakes2014; @Queiros2014] (reviewed by @Cahill2013 [@Cahill2013]), revealing that methylation patterns in CLL molecular subgroups have differing  methylomes that are likely an epigenetic inheritance of distinct normal B-cell subpopulations [@Kulis2012]. While DNA methylation has been shown capable of uncovering CLL groups with clinical significance [@Kulis2012; @Queiros2014], it is still a costly assay if completeness of coverage genome-wide is paramount. Additionally, DNA methylation is only one of the mechanisms which contribute to  control gene expression, with histone modification, chromatin accessibility and the nuclear architecture of the genome being important dimensions of the highly complex regulatory landscape of all cells.

Recent technological advances in chromatin profiling techniques, enable investigation into large patient cohorts [@Buenrostro2013] or rare cell populations by requiring less input material [@Lara-Astiaso2014a; @Schmidl2015], and bring great promise for large scale chromatin profiling in clinical settings. Since CLL shows substantial heterogeneity of molecular and clinical phenotypes (reviewed in by Gui and Wu [@Gui2015]), we decided to explore the amount of variation in its chromatin regulation. We performed the assay of transposase-accessible chromatin followed by next-generation sequencing (ATAC-seq) [@Buenrostro2013] in primary lymphocytes of 86 CLL samples from 58 patients, establishing a comprehensive map of epigenetic regulation which reveals the epigenetic variation in regulatory elements across the cohort. With a supervised machine learning approach that incorporates clinical annotations, we uncovered regulatory regions associated with the IGHV mutation status of groups of samples. Chromatin patterns at the most predictive regions reflect the common usage of regulatory regions of normal B cells with different maturation status and differentially open regions putatively regulate key genes in important signaling pathways, which might be important to the respective distinction between the more aggressive IGHV unmutated CLL, and the more indolent cases with mutated IGHV locus. By detecting transcription factor (TF) footprints in the ATAC-seq data, we infer gene regulatory networks (GRNs) of TF-gene interactions to gain deeper insights into the molecular wiring of CLL and describe the first GRNs of its kind for any cancer. GRN comparison between the IGHV molecular subgroups uncovers differentially used TFs with differential regulation of genes, which may shed light on the molecular basis of the groups of clinical phenotypes.

## Materials and methods
### Sample acquisition
All cases were diagnosed according to the revised guidelines of the International Workshop Chronic Lymphocytic Leukemia/National Cancer Institute (IWCLL/NCI).

IGHV sequencing, CD38 expression  and screening for CNA's were performed in all cases and ZAP70 expression in the majority. CNA's (del13q, del11q, TP53 loss and trisomy 12) were detected by FISH in most cases or MLPA (MRC Holland SALSA MLPA P037 CLL-1 probemix) in a minority. Most cases also had karyotypic analysis.

### ATAC-seq
Open chromatin mapping was performed using the ATAC-seq method as described [@Buenrostro2013] with minor adaptations for K562 cells. In each experiment, 10^5^ cells were washed once in 50 $mu$l PBS, resuspended in 50 $mu$l ATAC-seq lysis buffer (10 mM Tris-HCl, pH 7.4, 10 mM NaCl, 3 mM MgCl 2 and 0.01% IGEPAL CA-630) and centrifuged for 10 min at 4 $^{\circ}$C. Upon centrifugation, the pellet was washed briefly in 50 $mu$l MgCl 2 buffer (10 mM Tris, pH 8.0, and 5 mM MgCl 2 ) before incubating in the transposase reaction mix (12.5 $mu$l 2x TD buffer, 2 $mu$l transposase (Illumina) and 10.5 $mu$l nuclease-free water) for 30 min at 37 $^{\circ}$C. After DNA purification with the MinElute kit, 1 $mu$l of the eluted DNA was used in a qPCR reaction to estimate the optimum number of amplification cycles. Library amplification was followed by SPRI size selection to exclude fragments larger than 1,200 bp. DNA concentration was measured with a Qubit fluorometer (Life Technologies).
Sequencing was performed by the Biomedical Sequencing Facility at CeMM using the Illumina HiSeq 3000 platform. Library preparation was performed using custom Nextera primers as described previously [@Buenrostro2013].

### ATAC-seq data preprocessing
Reads were trimmed using Skewer [@Jiang2014] and the Nextera adapters as input sequences. Trimmed reads were aligned to the GRCh37/hg19 assembly of the human genome using Bowtie2 [@Langmead2012] with the "--very-sensitive" parameter and the aligner’s default. Duplicate reads were removed using sambamba *markdup*, and only properly paired reads with mapping quality over 30 aligning to the nuclear genome were kept. All downstream analyses were performed on these set of filtered reads.

Genome browser tracks were created with the *genomeCoverageBed* command in BEDTools [@Quinlan2010] and normalized such that each value represents the read count per base pair per thousand filtered reads. Finally, the UCSC Genome Browser's *bedGraphToBigWig* tool was used to produce a bigWig file. The tracks with percentile signal across the cohort was created by measuring the ATAC-seq read coverage at every basepair with BEDTools *coverage* and normalizing it between samples by divinding each position by the total number of filtered reads over 10 million. Then, the mean as well as the 5, 25, 75 and 95 percentiles of signal across the whole cohort were calculated with Numpy, made into bedgraph and subsequently bigwig format using UCSC's *bedGraphToBigWig*.

Peak-calling was performed with MACS2 [@Zhang2008] using the "--nomodel" and "--extsize 147" flags and arguments, and peaks overlaping the list of blacklisted features of the GRCh37/hg19 assembly as defined by the ENCODE project [@Hoffman2013a] were discarded.
<!--
Reads aligning to the plus strand were offset by +4 bp, and reads aligning to the minus strand were offset by -5 bp as described [@Buenrostro2013].
-->

### Analysis of chromatin accessibility variation
The CLL cohort ATAC-seq region set was created by merging all peaks across all samples using BEDTools [@Quinlan2010] *merge* command. To produce Figure [#Figure1]b, we counted the number of unique sites after merging peaks for each sample in an iterative fashion by randomizing the sample order 1000 times and computing 95% confidence intervals across all iterations.

We quantified the acessibility of each region over each sample using Pysam [@citeulike:8274107] by counting the number of reads from the filtered bam file overlaping each region. To normalize read counts across samples, we performed quantile normalization using the *normalize.quantiles* function from the preprocessCore package in *R*.

For each genomic region we calculated how much support it has over the cohort as the fraction of samples with a called peak in the region over all samples; five measures of ATAC-seq signal variation across the cohort: mean signal, standard deviation, dispersion (variance over mean) and the squared quoeficient of variation (the square of the standard deviation over the mean). Additionaly, we annotated each region with the identity of and distance to the nearest TSS of Ensembl gene models, and the type of genomic region overlaping by using BEDTools *intersect* with features (introns and exons) from the same gene annotation and considering promoters to be 2.5 kb upstream of TSSs and intergenic space the remaining space. To annotate each region with the chromatin state of CD19+ cells, we took the 15 state genome segmentation from the Roadmap Epigenomics Project [@Ernst2015] \(ID E032).

Gene-centric variation presented in Figure [#Figure2]a and b, represents the values of closest region the within 1 kb for promoters, and the mean signall of all distal regions away from the promoter (>1 kb). For Figure [#Figure2]a we ploted the intensity range for these regions, defined as the absolute difference between the 5th and 95th percentiles, whereas in Figure [#Figure2]b the intensity values of the respective regions are used directly to produce the violin plots.

### Sample classification on IGHV mutation status and region characterization
A Random Forest classifier from the scikit-learn [@Pedregosa2011] implementation (*sklearn.ensemble.RandomForestClassifier*) was trained with each sample's IGHV mutation status as label if known and the matrix of 112168 ATAC-seq CLL cohort region set per sample as input. All CLL samples were used for class prediction using leave-one-out cross validation, and a ROC curve was plotted using the scikit-learn's implementation. Important regions to distinguish samples with different IGHV mutation status were selected by averaging the feature importances of the Random Forest classifier over all iterations and selecting all features with non-zero importance.

Region set enrichment was performed using LOLA [@Sheffield2015] and its core databases: TF binding sites from ENCODE, tissue clustered DNase hypersensitive sites [@Sheffield2013], the Codex database, UCSC feature tables, the Cistrome database and data from the BLUEPRINT project.

Gene ontology and pathway enrichment of the IGHV-specific regions was performed using the R seq2pathway [@Wang2015] and TF motif enrichment analysis was performed with MEME's *ame* tool, using MEME's database of human TF motifs (HOCOMOCOv9).

The sample stratification in Figure [#Figure3]g was performed by correlating the ATAC-seq signal of each sample in the above-defined IGHV status regions of importance in a pairwise fashion and plotting a dendrogram of the correlation with Scipy's hierarchical clustering implementation. With the same correlation values, was performed with R's implementation.

### Gene regulatory network inference from TF footprints
Footprinting was performed with PIQ [@Sherwood2014] using a set of 366 Human transcription factor motifs from the Jaspar database [@Mathelier2014].

As described previously [@Qu2015] we chose to retain TFs with at least 500 high purity binding sites (> 0.7) that overlap any ATAC-seq regions of the CLL cohort.

Our scheme for assigning transcription factor binding sites to genes was as follows: TFBSs located in the body of a transcription unit or the 2.5 kb region upstream its transcription start site (TSS), were assigned to the respective overlaping gene - in the case of overlapping transcriptional units we assigned the TFBS to all overlaping genes. Intergenic TFBSs were assigned to the gene whose TSS was closest to the peak. For this association, we used the GRCh37/hg19 Ensembl gene annotation and we considered non-protein coding genes in the same manner as the protein coding ones.

<!--
If we need to justify this:
    - Most ATAC-seq peaks are within gene bodies:
    "transcriptionally active promoters asymmetrically interact with the gene body more than with local upstream sequence, suggesting an activating role for intronic enhancers" [Mifsud, B. et al. Mapping long-range promoter contacts in human cells with high-resolution capture Hi-C. Nat. Genet. 47, 598–606 (2015)]
    - Similar approaches have been used:
    [@Gonzalez2015] McLean, C.Y. et al. GREAT improves functional interpretation of cis-regulatory regions. Nat. Biotechnol. 28, 495–501 (2010).
-->

To infer gene regulatory networks, we calculated an interaction score in a similar way as previously done [@Qu2015]: the interaction score between a transcription factor $t$ and a gene $g$ ($S_{t,g}$) is given by the sum of all TF binding sites (of length $n$) from TF $t$ assigned to gene $g$:
~~~math #interaction-score
S_{t,g} = \sum_{i=0}^{n}2 * (P_{i} - 0.5) * 10 ^{-(d_{i, g})}
~~~
where $P$ is the PIQ purity score and $d_{i, g}$ is the distance of a particular TF binding site $i$ to gene $g$ in megabases. This establishes a unidirectional (TF to gene), weigthed (based on the interaction score) relationship, which forms the edges of a graph.

We infered gene regulatory networks for all samples and for groups of samples depending on its IGHV mutation stats if known. In the case of the two later cases, this was performed by using a concatenation of all of the samples’s bam files from each group as input to PIQ. We considered only TF-gene interactions with score higher than 1, and to produce Figure [#Figure4]b and Supplementary Figures 9 and 1] nodes with more than 200 connections. In order to compare the infered IGHV-unmutated and IGHV-mutated networks, we divided the degree of each node by the total number of edges in each network to compensate for differences in the absolute number of detected interactions and calculated change by subtracting and log2-transforming this value between networks for each node.

The entirety of the source code used in the analysis is available at [github.com/epigen/cll-patients](https://github.com/epigen/cll-patients).

## Results
### The chromatin accessibility landscape of CLL
To establish the chromatin accessibility landscape of CLL and investigate its variation across disease subtypes and particular cases, we used the assay of transposase-accessible chromatin followed by next-generation sequencing (ATAC-seq) [@Buenrostro2013] to generate patient-specific chromatin accessibility profiles (Figure [#Figure1]a).
We sequenced an average of 25.3 million read pairs per sample - a total of 2.25 billion read pairs across 86 samples (Supplementary Figure 1 and Figure [#Figure1]b), from sequencing libraries of high quality (Supplementary Figure 2). These were sequenced with enough depth to discover most accessibility regions in each sample (Supplementary Figure 3). With 86 CLL samples across the whole clinical range of CLL (see Supplementary Figure 4), we manage to cumulatively capture almost all unique sites of open chromatin (Figure [#Figure1]b).

To establish a defined region set that characterizes chromatin accessibility in CLL, we combined individual ATAC-seq peaks of all samples, creating a set of 112168 unique regions. A representative gene locus with the mean signal over all samples is displayed in Figure [#Figure1]c). Overall, many sites of the unique set were constitutively open in the cohort, whereas a considerable number of sites of open chromatin are either open in an exclusive set of samples or are private to a small number of samples (Supplementary Figure 5a).

CLL open chromatin regions are located mostly in intergenic space, gene introns and promoters (Figure [#Figure1]d and Supplementary Figure 5b), and near the transcription start sites of genes (and Supplementary Figure 5c), which matches what has been described for ATAC-seq data [@Buenrostro2013]. We decided to investigate the chromatin state of this set of regions in relation to CD19+ B cell populations (Figure [#Figure1]e and Supplementary Figure 5d). Consistent with the overall pattern of their genomic location, most peaks are located in known enhancers and transcription start sites (TSSs) of active genes in B cells. We do find, however, a considerable absolute amount of regions overlapping quiescent chromatin (without any enriched histone modification in particular) and transcriptional domains of B cells, although this amount is not particularly enriched when taking into account the distribution of the chromatin states in the genome (Supplementary Figure 5d).

### Heterogeneity in the CLL chromatin accessibility landscape
To investigate the heterogeneity of chromatin accessibility across the unique set of regions derived from the CLL cohort we quantified the accessibility level of each region in each sample. The average accessibility of all elements follows a skewed distribution (Supplementary Figure 6a), while inspection of variability in accessibility values across the cohort reveals a distribution with a group of regions with low dispersion, a group with intermediate, and a long tail of regions with high dispersion.
Since chromatin accessibility influences gene expression [@Natarajan2012; @Marstrand2014], we sought to characterize the variation in the chromatin accessibility of regulatory elements associated with genes. To this end, we associated each element with the likely gene it is regulating (see Methods) and quantified the dynamic range of accessibility in the promoter or putative distal regulatory elements of each gene by measuring the difference between the extremes of the distribution (5th to 95th percentiles) across all CLL samples. 
The range of chromatin accessibility values in promoter elements tends to be lower than that of distal regulatory elements of genes (Figure [#Figure2]a), although a considerable number of these also show high dynamic range2. Most distal regulatory elements of genes tend to exhibit similar variation across the CLL cohort (Figure [#Figure2]a).
To further describe the chromatin heterogeneity of CLL, we investigated the variability of regulatory elements of genes relevant to B cell biology and CLL pathogenesis across the cohort (Figure [#Figure2]b and Supplementary Figure 7a). 
Genes encoding B cell surface markers (CD19), BCR signaling components (CD79A/B, LYN, BTK), oncogenes (MYCN, KRAS, NRAS) or genes recursively mutated in CLL (NOTCH1, SF3BP1, XPO1, CDKN1B) [@Puente2015; @Landau2015] displayed a wide variety of both chromatin accessibility and dynamic range across the cohort. Perhaps of notice are the case of MYC and NRAS oncogenes, which both display high chromatin accessibility and low dynamic range across samples (Figure [#Figure2]b).

To assess if this set of CLL-relevant genes (n = 81, Supplementary Figure 7a) displayed a particular pattern in the variation of chromatin accessibility, we compared their dynamic range of accessibility with that of all genes. This gene set occupies mainly the lowest quantile of the distribution of chromatin accessibility of all genes (Supplementary Figure 7b), with distal elements being more dispersed through the quantiles. This apparent lower variability in key CLL genes could indicate that more consistent chromatin accessibility contributes to the overall CLL phenotype.

Nonetheless, some genes important to general B-cell development or B cell progression into more terminally differentiated states (e.g. memory B cells) such as the transcription factors PAX5 and BCL6 have a larger dynamic range of chromatin accessibility at both promoters and distal regulatory elements (Figure [#Figure2]c).

### Alternative chromatin use of CLL IGHV subtypes uncovered by supervised machine learning
We next sought to explore recurrent differences in chromatin accessibility between groups of CLL cases. To this end we used all 112168 CLL regions for a supervised machine learning classification guided by clinical annotation of CLL cases (Figure [#Figure3]a). This approach has the advantage that focusing clinically annotated features, increases the odds of finding clinically relevant information.
Since mutation of IGHV locus is the most determinant factor of CLL molecular variation and clinical outcome 1, we decided to focus on discovering chromatin accessibility regions associated with this feature. We trained a Random Forest classifier on the IGHV mutation status of the samples, and by using leave-one-out cross validation we predicted the IGHV mutation status of every sample, also if this had not been known or measured before. The classifier’s prediction accuracy was extremely high (AUC = 0.97 - Figure [#Figure3]b) even though we did not restrict the classification to any particular subset of regions, but used the combined set of CLL accessible chromatin regions across the whole cohort. 

Another advantage of this approach was that this learning process allowed us to easily extract features that are important to distinguish these two molecular subgroups without any hard cut-off. We found 1504 regions contributing to this distinction (Figure [#Figure3]c), which separated into two groups upon hierarchical clustering of the chromatin accessibility values. This duality is expected since IGHV mutation status was the trait guiding the learning process leading to the discovery of these regions.
Compared with the set of all CLL regions – from which these IGHV-associated regions are drawn – both the dynamic range and dispersion of chromatin accessibility values are increased, but not the support (Supplementary Figure 8a). Also importantly, there were no strong deviations from expected distributions regarding the chromosomal location of these regions (Supplementary Figure 8b) rejecting that any existing chromosomal aberrations between samples of different IGHV status would provide major differences in the location of these IGHV-associated regions.

To thoroughly characterise these regions, we separated them into two clusters: cluster 1, in which mCLL cases display higher accessibility (n = 719), and cluster 2 with higher uCLL accessibility (n = 764). We further investigated their genomic locations in context of the whole set of CLL chromatin accessibility regions. There are opposing patterns between the two groups of regions, especially in promoters and intergenic space (Figure [#Figure3]d). Promoters and exons were enriched in uCLL regions while promoters are depleted in the open chromatin regions associated with mCLL (Figure [#Figure3]d). A similar pattern was observed in relation to chromatin states of CD19+ B cells: uCLL regions are enriched in transcription start sites (TSS), while the mCLL regions are enriched in distal and intronic enhancers of CD19+ cells. While there was agreement in both clusters of regions in their depletion of repressive chromatin domains, perhaps the most striking finding is the divergence of uCLL and mCLL regions on bivalent TSSs and enhancers of CD19+ cells. These regions could be enhancers of progenitors that are poised for use by B cells during their maturation into plasma or memory B cells, or regions being differentially used in specific B cell populations, since whole CD19+ cells are a mixture of B cells in various maturation stages.

We therefore turned to a more functional characterization of these clusters of chromatin regions by measuring the overlap with collections of publicly available region sets of functional relevance including genomic and regulatory elements. Regions with higher chromatin accessibility in the uCLL were shown to be enriched in CLL-specific enhancer regions, and regions of CD38-negative naive B cells with the enhancer mark H3K4me1, reflecting the likely naive B cell origin of these CLL cells (Figure [#Figure3]e). These regions were also enriched in transcriptional domains (determined by ChIP-seq of H3K36me3) of naive B cells and B cell-derived cell lines such as the pre CSR BL-2 line, hinting that uCLL might be using new, possibly exclusive enhancers from the transcription sites of B cells. The enrichment of either DNase hypersensitivity regions or transcriptional domains of other hematopoietic cell types likely reflects the more undifferentiated status of uCLL in the B cell lineage compared to mCLL. On the other hand, regions more accessible in mCLL are dominated by enhancer regions of lymphocyte-derived cell lines (SU-DHL-5, JVM-2, GM12878, KARPAS-422) and TF binding sites (Figure [#Figure3]e). Examples of TF binding sites are BATF - known to regulate class-switch recombination (CSR) [@Ise2011], BCL6 – a master transcription factor for germinal center (GC) development, as well as EBF1, but also BCL3 and 11A - both associated with B-cell malignancies.

With cellular signaling pathways being a very important component of cellular activity and their activation often a determinant of proliferation and clinical outcome in cancer, we performed pathway enrichment of genes associated with IGHV-specific chromatin accessibility regions (Figure [#Figure3]f). We found striking differences between the two region clusters in terms of their enriched molecular pathways: the mCLL region cluster was enriched in pathways relevant for lymphocytes, but perhaps not particularly prominent in CLL pathogenesis (CTLA4 inhibitory signaling, the high-affinity IgE receptor, Fc$\epsilon$RI and the Fc receptor pathways) with the exception of BCR signaling; in the uCLL-specific regions we find NOTCH pathway signaling prominently enriched, as well as FGF signaling, together with several pathways that promote cytoskeletal activity and cell movement.
The sample structure created by hierarchical clustering of all samples in all of the 1504 IGHV-specific regions (Figure [#Figure3]c) clusters cases with the same IGHV mutation status together. Nonetheless, a group of samples seems to display enrichment of chromatin accessibility in an intermediate degree between the uCLL and mCLL groups suggesting that the sample structure is not hermetic but heterogeneous, and that further molecular groups of the disease may exist. To further structure CLL cases in the previously discovered IGHV-specific regions (Figure [#Figure3]c), we performed hierarchical clustering on the correlation of chromatin accessibility signal (Figure [#Figure3]g), as well as principal component analysis (PCA) on the chromatin accessibility signal (Figure [#Figure3]h) in these regions. The sample structure reveals three major groups, corresponding to uCLL and mCLL and a third intermediate group (iCLL) stemming mostly from the mCLL group, but having a mixture of IGHV mutation status. Interestingly, another intermediate group of samples stemming from the uCLL group is observed, with most samples bearing a mutated IGHV locus, despite its very small size (n = 3).

### Gene regulatory network inference from TF footprints
In the ATAC-seq assay, proteins binding to DNA can hinder adapter integration by the transposase enzyme, which locally reduces the accessibility signal. This protection, also called transcription factor footprints, can be detected through analysis of the accessibility signal in specific genomic positions containing TF motifs [@Buenrostro2013] by assessing the frequency of transposition events at each site (Figure [#Figure4]a, left). We applied transcription factor footprint detection using a set of 366 human transcription factor motifs to infer their binding to all CLL open chromatin regions. By assigning the bound locations to nearby genes we established a putative interaction between the TF and the gene in proximity of the binding site through an interaction score (see Methods), which we use to reconstruct gene regulatory networks (GRN) (Figure [#Figure4]a, right).

We first started with an attempt to reconstruct a GRN representing all TF-gene interactions mediated by TFs in CLL, by inferring the binding of all detectable TFs from the accessibility profiles of all CLL samples together (Supplementary Figure 9). In this composite CLL GRN, TFs are the most well-connected nodes4, among which some general regulators (SP1/2, CTCF), but also families of prominent regulators of proliferation (EGR, KLF) and cell cycle (E2F), altough some factors of relevance for B cell maturation (SPI1, PAX5) can still be seen among the top regulators of the CLL GRN (Supplementary Figure 9).

We reasoned that the same approach could be used to infer GRNs of specific CLL molecular subtypes, by considering regulatory interactions predicted from TF footprints in groups of samples. To gain deeper insights into the molecular wiring of CLL groups with different IGHV mutation status, we inferred GRNs specific to each group independently (Supplementary Figure 1]). Although the overall structure of the networks is similar, when comparing the node connectivity between the IGHV-specific networks, we find a range of genes which interaction degree is changing between the two networks (Figure [#Figure4]c and d). Among the genes with are most well connected in uCLL compared to mCLL are membrane molecules CD9 and CD38, the PAX9 transcription factor, the BMP receptor CRIM1 and the FGF signaling receptor FGFR1. The top specific regulators in mCLL include the ZNF354C transcription factor, the metaloprotease ADAM29, and the membrane receptor CD22, which functions as an inhibitory receptor for B cell receptor (BCR) signaling. We again investigated in more detail a set of genes related with B cell development, maturation, signaling as well as recursively mutated genes in CLL (Figure [#Figure4]d). We found that genes such as CD38 and ZAP70, which are known to be more characteristic of uCLL are among the set of more regulated genes in the uCLL GRN when compared with the mCLL. Conversely, CD22 is the top connected known-candidate gene in the cases of mCLL. The density of regulatory interactions in the neighbourhood of these nodes can be quite different between uCLL and mCLL, as illustrated for the case of PAX9 and CD22 (Figure [#Figure4]e). We also found a number of interactions exclusive to one of the groups. Although some of these differences could be explained by sensitivity differences arising by the different number of samples contributing to infer TF binding in each network, a number of genes containing a considerable number of interactions in one network (e.g. > 10) were not part of the other, due to the absence of any detected interaction. A representative example is the FGFR1 gene, in which we detected 13 TF interactions in uCLL but none in mCLL. This and other differentially regulated genes have noticeable variation in chromatin accessibility (Supplementary Figure 1]) consistent with the differential accessibility and likely, differential TF binding in their regulatory elements.

## Discussion
By performing ATAC-seq on a large set of primary CLL samples, we have for the first time characterized the chromatin accessibility landscape of CLL across the whole range of clinical subtypes of the disease. With this approach, we could reach saturation of chromatin accessibility sites in the genome in a relatively cost-effective manner, highlighting that chromatin accessibility profiling in a large cohort of cancer samples directly from primary patient samples is feasible, and likely to become more prevalent. The existence of a considerable amount of regions outside the expected context of promoters and known enhancers of whole B cell populations might indicate that these are either novel, CLL-specific enhancers or enhancers of more specific cell populations throughout B cell differentiation. Such a question would be answered by referring our dataset to chromatin profiles of B cell populations through their differentiation course.

With this project, we aimed to investigate the variation in chromatin regulation across the whole clinical range of 56 CLL patients. We've shown that variation in chromatin accessibility is widespread both at the level of single regulatory elements but also when considering all putative regulatory elements of genes, revealing the rather heterogeneous and likely dynamic chromatin landscape of CLL. The finding that genes relevant in the context of CLL (such as B cell-relevant pathways) display lower variability in the accessibility of their regulatory elements may suggest that active function of these genes is likely a requirement to achieve the molecular phenotype leading to the manifestation of CLL. However, the relative contribution of variation in chromatin accessibility to explain variation in gene expression remains to be elucidated. Additionally, signal arising from multiple CLL clones is confounded and this factor could contribute to intra-sample variation in chromatin accessibility. The recent development of single cell chromatin accessibility profiling (scATAC-seq)[@Buenrostro2015] could provide insights on the inter-clonal variation of chromatin accessibility in CLL.

The identification of IGHV-specific regions highlighted the genomic and epigenetic differences between the two major molecular subtypes of CLL: uCLL and mCLL. The functional investigation into these differentially used regions revealed that these likely reflect the potential cell-of-origin of the CLL cells: the mCLL group was enriched in enhancers of B cell lines and in TF binding sites of known factors required for differentiation into memory B cells (BATF, BCL6), while the uCLL group was enriched in regions used by naïve B cells and other hematopoietic cells, likely reflecting the more undifferentiated state of these cells. A common for both groups is the enrichment in regions used specifically for transcription (with the histone mark H3K36me3) in other cell types. We speculate that these are likely novel, CLL-exclusive enhancer regions, a hypothesis that would be clarified by histone mark profiling in the different CLL groups and normal B cell populations. The enriched signaling pathways between the two CLL groups may lead to new discoveries on the molecular phenotypes of these disease groups: inhibitory signaling by CTLA4 is a known pathway in T cells which leads to overall decrease in T cell population expansion [@Kong2014]. However, when CTLA4 was experimentally downregulated in CLL it showed increased cell survival through the upregulation of molecules involved in B-cell proliferation or survival signaling [@Mittal2013]. The fact that this pathway is exclusively enriched in the mCLL group could contribute to the generally less aggressive outcome of mCLL compared with uCLL. Signaling mediated through the Fc region of immunoglobulin E (FCERI) includes prominent BCR signaling molecules such as the prominent BCR signaling kinases LYN, SYK and BTK that may lead to NFKB activation. This contribution of Ca+ mediated FCERI signaling in CLL seems largely unknown and could therefore be a potential new area of molecular enquiry in the mCLL group. Also in mCLL, FC gamma receptor (FCGR) activation seems interesting: this molecule binds to the Fc portion of immunoglobulin G (IgG) and it is particularly interesting due to its relation with CD20 - a common clinical target in lymphoid malignancies: high FCGR expression reduces Rituximab efficiency by causing CD20  internalization [@Lim2011], and FCGR polymorphisms have been shown to alter the clinical efficacy of rituximab [@Zhuang2015]. Enrichment of the BCR signaling pathway in mCLL was surprising, particularly in light of its higher activity in uCLL. Nonetheless, antigen-independent cell-autonomous BCR stimulation is also known to broadly occur in CLL [@Minden2012]. The sample substructure we found when measuring chromatin accessibility in IGHV-specific regions is similar to that previously found in DNA methylation studies of CLL [@Kulis2012; @Queiros2014], with a third group of CLL cases (iCLL) which seems to stem primarily from the mCLL group, having a mixture of the chromatin accessibility signature of the uCLL and mCLL groups in these regions. The existence of a fourth group, of iCLL more close to the uCLL one would be of interest and if expanded upon the analysis of more samples, would certain warrant further molecular investigation of these cases. Supervised machine-learning approaches based on clinical annotations seems therefore valuable in uncovering sample substructure and characterization of regions contributing to the distinction between sample groups and could be more broadly applied to new groups of cases with features of clinical relevance (e.g. different therapies, overall survival).

Although reverse engineering of regulatory networks in human cells has been a goal for a while (inclusively in B cells [@Basso2005; @Lefebvre2010]) large-scale detection of regulatory interactions in a patient-specific manner has never been achieved. This work presents thus the first application of GRN inference through TF footprinting in a cancer setting cancer. While the network infered from all CLL samples might reveal general cellular activity or pan-cancer signals, some B cell-specific TFs are also shown to have high regulatory activity. Such examples include important factors for B cell maturation such as BCL6, PAX5 and SPI1, but other factors known to be markers of CLL progression such as E2F4 [@Falt2005]. Among the most differentially regulated TFs between uCLL and mCLL, we have detected a few known cases that have IGHV mutation status-dependent regulation: several genes in the HOXA cluster are among the most differentialy regulated genes, although interestingly, in opposing directions. Since this locus has been shown to be hypermethylated in mCLL [@Strathdee2006] we hypotesise that DNA methylation at its locus in mCLL and prevents the binding of TFs either directly or by recruitment of chromatin modifiers which repress the locus' chromatin. Inference of GRNs through TF foootprinting seems therefore capable of retrieving known regulatory interactions in disease subgroups and due to the genome-wide nature nature of ATAC-seq data, is likely to reveal many new interactions of potential clinical relevance. This approach should be of particular use in cancer, where where the gene regulatory program of cells is often changed, but also in percision medicine in the context of several diseases since it is possible to infer GRNs with relative confidence albeit with low sensitivity from a single ATAC-seq sample.

This work systematically dissects the regulatory patterns and inter-patient heterogeneity of CLL based on chromatin data for a large patient cohort – finding groups of molecular phenotypes with distinct usage of regulatory regions and signalling pathways, and specific regulatory interactions of potential therapeutic relevance.

## Acknowledgements
We thank the Biomedical Sequencing Facility at CeMM for assistance with next-generation sequencing and all members of the Bock lab for their help and advice. This work was performed in the context of the BLUEPRINT project (European Union’s Seventh Framework Programme grant agreement no. 282510) and funded in part by the ERA-NET project CINOCA (FWF grant agreement no. I 1626-B22). C.S. was supported by a Feodor Lynen Fellowship of the Alexander von Humbold
t Foundation. C.B. was supported by a New Frontiers Group award of the Austrian Academy of Sciences.


## Authorship
**Contribution:**
R.W. and Z.D. followed the patients and isolated lymphocytes from periferal blood, D.O. and J.S. contributed the samples, C.S. performed the experiments; A.F.R. analyzed the data; A.F.R., C.S., D.O., J.S. and C.B. planned the study; C.B. supervised the research; all authors contributed to the writing of the manuscript.

**Conflict-of-interest disclosure:**
The authors declare no competing financial interests.

**Correspondence:**
Christoph Bock, CeMM Research Center for Molecular Medicine of the Austrian Academy of Sciences, Vienna, Austria; e-mail: <cbock@cemm.oeaw.ac.at>.


## Tables
## Figures
### Figure: {#Figure1}
![](figures/Figure1.pdf){height=100%}
Caption: *The chromatin accessibility landscape of chronic lymphocytic leukemia.*
**a)** Schematic representation of the experimental procedures (top) leading to a map of chromatin accessibility on a CLL cohort.
**b)** Cumulative frequency of unique sites of open chromatin detected across CLL samples in the entire cohort and total amount of read pairs sequenced. Lines indicate 95% confidence interval.
**c)** Representative illustration of ATAC-seq signal in individual CLL samples and the average signal across the whole cohort in the BLK locus.
**d)** Enrichment of genomic regions across all detected regionse) Enrichment of chromatin states of CD19+ cells across all detected regions in the cohort.

### Figure: {#Figure2}
![](figures/Figure2.pdf){height=100%}
Caption: *Inter-sample variation in the CLL chromatin accessibility landscape.*
**a)** Variation of accessibilty in promoter and distal regulatory element for all genes with assigned accessible regions. Intensity for genes with more than one distal regulatory element was averaged across elements. Intensity range is defined as the absolute difference between the 95% and 5% percentile of ATAC-seq signal in each element across the cohort.
**b)** Violin plots of chromatin accessibility measurements (openness) at the promoter and enhancers (as defined above) of genes relevant to B cell biology.
**c)** ATAC-seq signal across the CLL cohort in examplary loci of genes in panel **b)**.

### Figure: {#Figure3}
![](figures/Figure3.pdf){height=100%}
Caption: *Clinical data-driven analysis of chromatin accessibility provides insights into CLL diversity and biology.*
**a)** Schematic representation of the machine learning aproach based on clinical data (IGHV mutation status in this case) combined with chromatin accessibility data to achieve sample/patient stratification and relevant feature extraction.
**b)** Receiving operating characteristic (ROC) curve of binary sample classification using a random forest classifier trained on 112168 regions and the IGHV mutation status of the samples. The ROC curve was produced after leave-one-out cross validation and the area under the curve is displayed (0.97).
**c)** Heatmap of samples and regions important to distinguish between the status of IGHV mutation in CLL samples during the classification exercise. Rows and columns were clustered hierarchicaly. Note the two clusters of regions each enriched in IGHV unmutated or mutated samples and therefore termed (uCLL or mCLL, respectively).
**d)** Enrichment of the clusters of regions defined in **c)** in the chromatin states of CD19+ B cells over all CLL open chromatin regions across the cohort.
**e)** Enrichment of cluster 1 and 2 (as defined in **c)**) in predefined functional genomic region sets from other cell types or biological conditions using LOLA.
**f)** Enrichment in cellular pathways from the Reactome and KEGG databases of cluster 1 and 2 (as defined in **c)**)-associated genes.
**g)** Hierarchical clustering of CLL samples based on the sample-wise correlation of chromatin accessibility values of all regions defined in **c)**. Samples are colored with the status of IGHV mutation and with the cluster in which they are placed in the dendrogram.
**h)** Principle component analysis of CLL samples with the chromatin accessibility values of all regions defined in **c)**. Sample colors are as defined in **g)**.

### Figure: {#Figure4}
![](figures/Figure4.pdf){height=100%}
Caption: *Gene regulatory network inference from transcription factor footprints in chromatin accessibility data.*
**a)** Schematic representation of transcription factor (TF) footprinting, TF-gene interaction assignment and gene regulatory network inference from such interactions.
**b)** Genes most differentially regulated between gene regulatory networks inferred from TF footprinting of uCLL or mCLL samples. Node size reflects the absolute degree (total number of connections) of the node and colour the network in which the node has more connections (orange for uCLL; blue for mCLL).
**c)** Logarithm of the difference in node degree between the uCLL and mCLL networks for all genes.
**d)** Logarithm of the difference in node degree between the uCLL and mCLL networks for a set of 81 CLL-relevant genes.
**e)** Subnetworks of interactors oftwo of the most differentially regulated genes from **c)**.


## Supplemental data
#### Supplementary Figure 1
Descriptive statistics on total number of sequenced, duplicate and filtered reads for each sample in this study.

#### Supplementary Figure 2
Distribution of sequenced fragment lengths from GM12878 and CD4+ T cells from Buenrostro et al12, and four randomly selected CLL samples from this study as a measure of ATAC-seq library quality. High quality libraries display characteristic distribution of fragment lengths.

#### Supplementary Figure 3
Sample-wise estimation of sequencing saturation based the number of novel region discovered with increasing sequencing depth at different q-value thresholds.

#### Supplementary Figure 4
Description of clinical attributes of samples used in this study.

#### Supplementary Figure 5
Genomic features of the CLL chromatin accessibility landscape. a) Support level of each region across the cohort. Support is defined as the fraction of samples containining an ATAC-seq peak over all samples. b) Frequency of each type of genomic region in the set of unique CLL chromatin accessibility regions and in a set of same size and feature lengths randomized across the genome 1000 times. Enrichment of CLL regions frequency over background is shown. c) Distribution of distances to the nearest transcription start site of each CLL chromatin accessibility region. d) Frequency of CD19+ chromatin states in the set of unique CLL chromatin accessibility regions and in a set of same size and feature lengths randomized across the genome 1000 times. Enrichment of CLL regions frequency over background is shown.

#### Supplementary Figure 6
Mean normalized accessibility values and dispersion (variance over mean) of each CLL region across the cohort.

#### Supplementary Figure 7
Heterogenity in chromatin accessibility in B-cell and CLL related genes. a) Violin plots of normalized accessibility values for the promoter (accessibility regions within 2 kb of the start site) or the mean of any distal regulatory element (at least 2 kb away from the start site) for 81 genes relevant for B cell biology or CLL. b) Distribution of the dynamic range (absolute difference between the 95th and 5th percentiles) in the quantiles of dynamic range of chromatin accessibility of all genes.

#### Supplementary Figure 8
Features of IGHV mutation statu-specific regions. a) Distribution of dynamic range, support and dispersion of all CLL regions and IGHV-specific regions  b) Frequency of chromosomal locations of the IGHV-specific regions.

#### Supplementary Figure 9
The CLL GRN. a) Gene regulatory network (GRN) of CLL infered from footprint predictions based on the concatenation of all CLL samples. Only nodes with degree (number of connections) higher than 200 are displayed. b) Degree of all genes in the CLL GRN.

#### Supplementary Figure 10
IGHV-specific CLL GRNs. Gene regulatory networks of IGHV unmutated (uCLL) and mutated (mCLL) samples. Only nodes with degree (number of connections) higher than 200 are displayed.

#### Supplementary Figure 11
Genomic loci of differentially regulated genes between the uCLL and mCLL gene regulatory networks.


### Tables


### Website
Track hub with consensus (6 tracks: percentile 5, 25, 50, 75, 95, mean) & all the individual tracks

## References
