---
title: Chromatin accessibility maps for chronic lymphocytic leukemia identify common regulatory patterns and patient-specific cancer heterogeneity
author:
        - André F. Rendeiro^1,\*^,
        - Christian Schmidl^1,\*^,
        - Jonathan Strefford^2,\*^,
        - Renata Walewska^3^,
        - Zadie Davis^3^,
        - Matthias Farlik^1^,
        - David Oscier^3^,
        - Christoph Bock^1,4,5,$\sharp$^,
        - ^1^CeMM Research Center for Molecular Medicine of the Austrian Academy of Sciences, Vienna, Austria
        - ^2^Cancer Genomics, Cancer Sciences, University of Southampton, Southampton, UK
        - ^3^Department of Molecular Pathology, Royal Bournemouth Hospital, Bournemouth, UK
        - ^4^Department of Laboratory Medicine, Medical University of Vienna, Vienna, Austria
        - ^5^Max Planck Institute for Informatics, Saarbrücken, Germany
        - ^\*^These authors contributed equally to this work
        - ^$\sharp$^Correspondence: <cbock@cemm.oeaw.ac.at>
tags: [Chronic lymphocytic leukemia, chromatin accessibility, epigenome mapping, cancer heterogeneity, leukemia, hematopoietic malignancies, transcription regulation, gene regulatory networks, bioinformatics]
bibliography: /home/afr/Documents/library.bib
---

## Abstract:
Chronic lymphocytic leukemia (CLL) is characterized by substantial clinical heterogeneity, despite its rather low burden of somatic mutations. To provide a broad basis for studying the role of chromatin regulation in CLL biology, we established genome-wide chromatin accessibility maps for 88 CLL samples from 55 patients using the ATAC-seq assay, and we devised bioinformatic methods for linking chromatin regulation to clinical data. The chromatin data identified a shared core of CLL-regulatory regions, but also widespread variation between disease subtypes and among individual patients. IGHV mutation status – which distinguishes the two major subtypes of CLL – was accurately predicted by chromatin state, and gene regulatory networks inferred for IGHV mutated vs. IGHV-unmutated samples identified characteristic regulatory differences between these two disease subtypes. In summary, we found widespread heterogeneity of chromatin accessibility in CLL, established a community resource for studying chromatin regulation and cellular heterogeneity, and demonstrated the feasibility of chromatin accessibility mapping in cancer cohorts as a research tool and potential clinical biomarker.

## Introduction
Chronic lymphocytic leukemia (CLL) is the most common type of leukemia in the Western world[@Byrd2004]. CLL is characterized by remarkable clinical heterogeneity, with some patients pursuing an indolent course while others require early treatment. The heterogeneity in the clinical course of CLL, particularly for those patients that present with a low disease burden, drives the continuing impetus for developing biomarkers with prognostic value[@Zenz2010]. Relevant clinical biomarkers in CLL currently include IGHV gene mutational status[@Damle1999; @Hamblin1999], IGHV gene family usage[@Tobin2002], stereotyped B cell receptors[@Agathangelidis2012; @Rossi2009], serum markers[@DiGiovanni1989; @Hallek1999], chromosomal aberrations[@Dohner2000; @Rossi2013], and somatic mutations[@Baliakas2014; @Oscier2012; @Stilgenbauer2014].

Whole genome and exome sequencing have identified several dozen genes that are recurrently mutated in CLL[@Puente2015; @Landau2015], some of which have prognostic significance. However, CLL samples carry relatively few somatic mutations compared to other adult cancers[@Lawrence2013], and some patients develop progressive disease despite being classified as “low risk” based on established biomarkers, suggesting that non-genetic factors may contribute significantly to CLL etiology and outcome. In particular, several lines of evidence suggest that chromatin accessibility and epigenome deregulation are relevant for CLL pathogenesis: First, recurrent somatic mutations have been observed in non-coding regions of the genome, where they appear to induce deregulation of relevant cancer genes[@Puente2015]. Second, chromatin remodeling proteins such as ARID1A and CHD2 are recurrently mutated in CLL[@Puente2015; @Landau2015], suggesting causal links between chromatin deregulation and CLL. Third, aberrant DNA methylation patterns were observed in all studied CLL patients[@Kulis2012; @Landau2014; @Oakes2014] and correlated with disease progression and distinguishing between CLL subtypes such as mCLL (with mutated IGHV) and uCLL (with unmutated IGHV). Fourth, DNA methylation analysis identified an intermediate form described as iCLL[@Kulis2012; @Queiros2014].

Recent technological advances make it possible to study chromatin accessibility and epigenome regulation in large cohorts of primary patient samples. Most notably, the assay for transposase-accessible chromatin using sequencing (ATAC-seq) provides a powerful method for open chromatin mapping throughout the genome[@Buenrostro2013], which is highly complementary to existing DNA methylation assays that focus on repressive chromatin. In ATAC-seq, the cellular chromatin is fragmented using a hyperactive variant of the prokaryotic Tn5 transposase, which integrates sequencing adapters preferentially into the DNA of genomic regions with accessible chromatin structure. ATAC-seq gives rise to globally similar results as DNase-seq, and both assays share the ability to detect footprints of transcription factor binding in the chromatin[@Risca2015]. Furthermore, the ATAC-seq assay can be performed accurately on scarce patient samples and on timescales that would be compatible with routine clinical diagnostics.

To establish the feasibility of large-scale chromatin analysis in primary cancer samples, and to provide a basis for dissecting regulatory heterogeneity in CLL, we performed chromatin accessibility mapping using the ATAC-seq assay on a cohort of 88 primary CLL samples. We also established bioinformatic methods for supervised and unsupervised analysis of the resulting data, and we gained initial insights into the widespread regulatory heterogeneity that underlies the major disease subtypes of CLL. In summary, this study provides a publicly available reference dataset and a rich source of testable hypotheses for dissecting CLL biology and pathogenesis.

## Results
### Chromatin accessibility maps for 88 CLL samples
To establish the chromatin accessibility landscape of CLL (Figure [#Figure1]a), we performed ATAC-seq22 on 88 primary disease samples obtained from 55 patients. All patients were managed at the same medical center, and they collectively represent the spectrum of clinical phenotypes that are commonly observed in CLL (Supplementary Table 1 and Supplementary Figure 1). 51 samples were IGHV mutated, whereas 37 samples were IGHV unmutated. All ATAC-seq libraries were subjected to paired-end sequencing with an average of 25.3 million sequenced fragments per sample, resulting in a dataset with a total of 2.25 billion sequenced fragments (Supplementary Table 2). Data quality was high in all cases, with mitochondrial and duplicate read rates in the expected range (Supplementary Table 2) and the characteristic patterns of nucleosome phasing (Supplementary Figure 2).

Each of the individual CLL samples was sequenced with sufficient depth to detect the majority of chromatin-accessible regions, with a clear trend toward saturation at the chosen sequencing depth (Supplementary Figure 3). Furthermore, by combining data for 88 samples across the whole clinical spectrum range of CLL we also approached cohort-level saturation in terms of unique chromatin-accessible regions (Figure [#Figure1]b). As illustrated for the BLK gene locus (Figure [#Figure1]c), our ATAC-seq dataset can be aggregated into a comprehensive map of chromatin accessibility in CLL. Our inferred chromatin accessibility landscape of CLL comprises 112,168 candidate regulatory regions, of which 11.6% are constitutively open across essentially all CLL samples, whereas 59.1% are open in a sizable proportion (5% to 95% of samples), and 29.3% are unique to only one or very few samples (Supplementary Figure 4a). All data are available for interactive browsing and download from the supplementary website (<http://rendeiro2016.computational-epigenetics.org/>).

Chromatin-accessible regions in CLL are widely distributed throughout the genome, with relative enrichment for genes and gene promoters (Figure [#Figure1]d and Supplementary Figure 4b). We also compared these CLL-accessible regions to epigenome segmentations of CD19+ B cells (Figure 1e and Supplementary Figure 4c), a related cell type for which comprehensive reference epigenome data are publicly available24. The CLL-accessible regions were highly enriched for transcription start sites and enhancer elements that are also active in CD19+ B cells. But we also observed a sizable number of CLL open chromatin regions that were associated with quiescent or repressive chromatin in the B cell data, suggesting that many of them may constitute CLL-specific regulatory elements.

### Heterogeneity in the CLL chromatin accessibility landscape
To investigate heterogeneity across samples, we calculated the variance in chromatin accessibility for each of the 112,168 regions in the consensus map, and we linked these regions to the nearby genes that they may regulate (see Methods for details). We observed significantly reduced variability ($p$ < 10^-5^) at the promoters of genes with a known role in B cell biology and CLL pathogenesis (Figure [#Figure2]a), suggesting that consistently high levels of chromatin accessibility at many of these gene promoters may be important for CLL cells. In contrast, distal enhancer elements were equally variable for genes with and without a biological link to CLL (*p* = 0.08).

Beyond these global trends, the variance and distribution of chromatin accessibility across samples was highly gene-specific (Figure [#Figure2]b and Supplementary Figure 5), which is illustrated by CLL-linked genes including B cell surface markers (CD19), BCR signaling components (CD79A/B, LYN, BTK), common oncogenes (MYCN, KRAS, NRAS) and genes that are recurrently mutated in CLL (NOTCH1, SF3BP1, XPO1, CDKN1B)15, 16. Notably, extensive variation across samples was observed for key regulators of B cell development and differentiation such as PAX5 and BCL6 (Figure 2c), suggesting that the observed variance in chromatin accessibility may have a direct effect on regulatory cell states. To facilitate the gene-by-gene investigation and future functional validation of these patterns, we established the “chromatin accessibility corridor” as a means of aggregating the cohort-level variation into a single intuitive genome browser track (Figure [#Figure2]c and Supplementary Website).

Linking the observed heterogeneity to clinical subtypes of CLL, we identified 389 CLL-accessible regions that show a higher degree of variation among mCLL samples, whereas 581 regions have higher variation among uCLL samples (Figure [#Figure2]d). These differentially variable regions are distributed across a broad range of chromatin accessibility (Figure [#Figure2]e), thus confirming that this observation is not a side effect of differences in the average chromatin accessibility between groups. To functionally annotate the differentially variable regions, we assessed their overlap with a catalog of annotated region sets using the LOLA software[@Sheffield2015] and found mCLL-variable regions enriched for B cell specific transcription factor binding (ATF2, BATF, BCL6, NFKB, RUNX3) and active histone marks (Figure [#Figure2]f). In contrast, uCLL-variable regions were strongly associated with the cohesin complex, including binding sites for CTCF, RAD21, and SMC3.

### Disease subtype-specific patterns of chromatin accessibility
For a more comprehensive analysis of CLL heterogeneity, we developed a machine learning approach that infers chromatin patterns predictive of clinical disease subtypes directly from the data (Figure [#Figure3]a). Focusing on IGHV mutation status, we trained random forest classifiers to predict whether a particular sample should be categorized as mCLL or uCLL, based on normalized chromatin accessibility for the 112,168 regions of the CLL consensus map. The performance of this classifier was evaluated by leave-one-out cross-validation, and we observed excellent prediction accuracy with a ROC area under curve of 0.97 (Figure [#Figure3]b), corresponding to a sensitivity of 95.9% at a specificity of 88.6%. This excellent performance confirms that the chromatin accessibility landscapes of mCLL and uCLL are characteristically different and reproducibly predict IGHV mutation status.

From the random forest classifiers we extracted the most predictive regions, resulting in a differential chromatin accessibility landscape of IGHV mutated vs. IGHV unmutated CLL. Hierarchical clustering categorized the predictive regions into 719 with increased chromatin accessibility in IGHV mutated samples (cluster 1 “mCLL regions” in Figure [#Figure3]c) and 764 regions with increased chromatin accessibility in IGHV unmutated samples (cluster 2 “uCLL regions” in Figure [#Figure3]c). Compared to all CLL chromatin-accessible regions, these regions were more variable across the cohort (Supplementary Figure 6a). In contrast, there was little chromosomal clustering and no evidence suggesting that the observed differences might be primarily associated with chromosomal aberrations (data not shown).

Among the mCLL regions we observed depletion of promoters and regions of repressed chromatin in CD19+ B cells, whereas uCLL regions were enriched for promoters, exons, and bivalent enhancers in CD19+ B cells (Figure [#Figure3]d). These results may indicate that IGHV unmutated CLL activates poised enhancers that confer cellular plasticity to B cells. We also annotated the two region sets using LOLA[@Sheffield2015] (Figure [#Figure3]e) and found that the mCLL regions were enriched for active promoter and enhancer regions (marked by the histone modifications H3K4me1 and H3K27ac) in lymphocyte-derived cell lines (SU-DHL-5, JVM-2, GM12878, and KARPAS-422) as well as binding sites of relevant transcription factors (BATF, BCL6, and BLC3). In contrast, the uCLL regions were enriched for H3K4me1-marked promoter/enhancer regions in CD38-negative naive B cells, reflecting the likely naive B cell origin of these CLL cells. The uCLL regions were also enriched for transcribed regions (histone H3K36me3) in naive B cells and in B cell-derived cell lines (such as the BL-2 cell line, which has not undergone class-switch recombination). 

Linking chromatin-accessible regions to genes (see Methods for details), we also investigated the enrichment of cellular signaling pathways and observed striking differences (Figure [#Figure3]f). The mCLL regions were associated with pathways having an established role in normal lymphocytes (CTLA4 inhibitory signaling, the high-affinity IgE receptor, Fc Epsilon (Fc$\epsilon$RI) and Gamma (Fc$\gamma$R) receptor pathways), whereas the uCLL regions were associated with more directly cancer-associated pathways such as NOTCH signaling and FGF receptor signaling.

Finally, we tested whether there was evidence for a third group of IGHV intermediate samples (iCLL) in our dataset, a new CLL subtype that has recently been proposed based on DNA methylation data18, 21. We clustered all samples by their chromatin accessibility across all IGHV-specific regions and indeed observed two iCLL clusters, the larger one comprising 20 samples from 13 patients (Figure [#Figure3]g, green) and the smaller one comprising 3 samples from 2 patients (Figure 3g, yellow). Most but not all of these iCLL samples were classified as IGHV-mutated in the clinical annotation (Supplementary Figure 6b). Principal component analysis confirmed that the chromatin accessibility landscapes of these two clusters were indeed intermediates between the mCLL and uCLL groups (Figure [#Figure3]h).

### Gene regulatory networks underlying the mCLL and uCLL disease subtypes
On top of chromatin accessibility mapping, ATAC-seq can also provide information on transcription factor binding. The bioinformatic approach is similar to DNase-seq footprint analysis22, and it allowed us to infer chromatin-based gene regulatory networks for CLL and its disease subtypes (Figure [#Figure4]a). To that end, we combined ATAC-seq data across all mCLL and uCLL samples, mapped footprints for 366 transcription factors with high-quality motifs in the JASPAR database26, and linked these regulatory elements to their putative target genes (see Methods for details of the gene regulatory network inference).

First, we inferred a pan-CLL gene regulatory network using the combined ATAC-seq data from all samples (Supplementary Figure 7a). The resulting network was dominated by highly connected transcription factors (Supplementary Figure 7b), including ubiquitous regulators such as SP1/2/3, the insulator protein CTCF, as well as regulators of cell proliferation (EGR, KLF), cell cycle (E2F), and B cell maturation (SPI1, PAX5).

Second, in order to investigate the regulatory differences between CLL subtypes, we inferred gene regulatory networks separately for mCLL and for uCLL samples. The global patterns were similar between the two subtype-specific networks (Supplementary Figure 8), but we also observed several interesting differences (Figure [#Figure4]b and c). Genes that were more highly connected in mCLL as opposed to uCLL included the transcription factors ZNF354C and ELF5, the metallopeptidase ADAM29, and the membrane protein CD22. In contrast, the BMP receptor CRIM1, the transcription factors MECOM and PAX9, the FGF signaling receptor FGFR1, and the membrane protein CD9 were more highly connected in uCLL.

When we restricted our analysis to genes with a known role in B cell biology and CLL pathogenesis (Figure [#Figure4]d), we observed a highly specific association of CD22 (an inhibitory receptor for B cell receptor signaling) with mCLL, whereas the CD38 and ZAP70 genes were preferentially associated with uCLL. Finally, focusing on CD22 and PAX9 as two interesting genes emerging from this analysis, we contrasted the sub-networks of their direct neighbors in the gene regulatory networks of mCLL and uCLL (Supplementary Figure 9), illustrating the characteristic regulatory differences between these two disease subtypes.

## Discussion
By performing ATAC-seq on a large set of primary CLL samples, we have established a comprehensive map of the chromatin accessibility landscape in CLL. To our knowledge, this dataset is the largest catalog of open chromatin maps of any single cancer type, which demonstrates the feasibility and relevance of ATAC-seq analysis in cohorts of primary cancer samples. Focusing our analysis of this dataset on cancer heterogeneity[@Alizadeh2015], we detected widespread variation in chromatin accessibility between patients but also a large core set of open chromatin regions that are shared across most or all CLL samples that we analyzed. The major disease subtypes of CLL – namely IGHV mutated mCLL and IGHV unmutated uCLL – were uniquely identified by their characteristic chromatin landscape, and we found evidence for additional disease subtypes that seem to be adequately described as intermediate forms of CLL (iCLL). Finally, for the mCLL and uCLL subtypes we collectively reached the required sequencing depth for footprint analysis of transcription factor binding, which allowed us to infer gene regulatory networks from the data and to compare them between the two major disease subtypes of CLL.

While our study cannot mechanistically prove that the observed chromatin heterogeneity is indeed causal for the widespread clinical heterogeneity of CLL, several lines of evidence are suggestive of functional links. First, we found widespread variation in chromatin accessibility for genes with a known role in B cell biology and CLL pathogenesis. This variation was lower than the genomic average for the promoters of these genes but high among enhancers associated with these genes. Second, chromatin accessibility maps turned out to be highly predictive of IGHV status, thus discriminating between mCLL and uCLL. This association appears to reflect differences in the cell-of-origin, which are conserved in the chromatin landscape and influence the molecular and clinical phenotypes of the disease subtypes. Specifically, mCLL-associated regions were enriched for mature B cell enhancers and transcription factors associated with differentiation into memory B cells (BATF, BCL6) whereas the uCLL group was enriched for regulatory regions that are active in other hematopoietic cell types, which is indicative of a less differentiated state of these cells. Third, the observed differences in pathway activity between mCLL and uCLL appear to be consistent with the generally more indolent course of mCLL. For example, inhibitory signaling by CTLA4 appears to be substantially reduced in uCLL, which may result in increased cell survival through the upregulation of molecules involved in B cell proliferation and survival signaling[@Mittal2013].

From a technological perspective, our study describes broadly applicable methods for dissecting chromatin patterns in large and clinically annotated cohorts of primary patient samples. The differential chromatin analysis outlined in Figure [#Figure3] starts from clinical data and uses supervised learning techniques to identify and cross-validate discriminatory chromatin signatures. Here we focused on IGHV mutation status, but the method can be applied to any type of patient grouping, for example based on age or therapy response. Moreover, the described method for ATAC-seq based inference of gene regulatory networks (Figure 4) provides a data-driven approach to dissecting regulatory cell states – including their differences between disease subgroups – that is highly complementary to previous work aimed at inferring regulatory networks from transcriptome data[@Basso2005; @Lefebvre2010; @Yepes2015]. Finally, the “chromatin accessibility corridor” provides an adaptation of a related concept[@Bock2011] and intuitive browser-based visualization for chromatin data across large sample cohorts with significant regulatory heterogeneity, which is expected to facilitate the re-use of the presented dataset and resource by non-bioinformaticians.

Relevant limitations of our study include: *i*) Lack of a clearly defined and experimentally accessible cell-of-origin for uCLL and mCLL, which makes it difficult to distinguish with certainty between chromatin patterns that are CLL-specific and those that are cell-of-origin derived; *ii*) clonal heterogeneity of CLL within patients, which would be experimentally addressable only with single-cell sequencing technologies[@Buenrostro2015; @Jin2015] that are currently limited in their genome-wide coverage and technically difficult for primary CLL samples; *iii*) lack of suitable methods for distinguishing between regulatory and spurious transcription factor binding; and *iv*) ambiguities in the assignment of transcription factor binding sites to the genes that they regulate. In the light of these limitations, the inferred gene regulatory networks constitute an initial model that will require future refinement as additional data and validations become available.

In summary, our study establishes a chromatin accessibility landscape of CLL and it identifies widespread regulatory heterogeneity. It also provides a resource that can act as a starting point for deeper dissection of chromatin regulation in CLL, identification of therapeutically relevant mechanisms, and the eventual translation of relevant discoveries into clinical practice – for example using chromatin accessibly as a biomarker for stratified therapy.

## Acknowledgements
We would like to thank all patients who have donated their samples for this study. Furthermore, we thank the Biomedical Sequencing Facility at CeMM for assistance with next-generation sequencing and all members of the Bock lab for their help and advice. This work was performed in the context of the BLUEPRINT project (European Union’s Seventh Framework Programme grant agreement no. 282510). C.S. was supported by a Feodor Lynen Fellowship of the Alexander von Humboldt Foundation. J.C.S. was supported by Bloodwise (11052, 12036), the Kay Kendall Leukaemia Fund (873), Cancer Research UK (C34999/A18087, ECMC C24563/A15581), Wessex Medical Research, and the Bournemouth Leukaemia Fund. C.B. was supported by a New Frontiers Group award of the Austrian Academy of Sciences.

## Author contributions
A.F.R., C.S., J.S., D.O., and C.B. planned the study; J.S., R.W., Z.D., and D.O. provided samples and clinical data; C.S. and M.F. performed the experiments; A.F.R. analyzed the data with contributions from C.S., J.S., D.O., and C.B.; C.B. supervised the research; all authors contributed to the writing of the manuscript.

## Competing financial interests
The authors declare no competing financial interests.

## Figures
### Figure: {#Figure1}
![](figures/Figure1.pdf){width=100%}
Caption: *The chromatin accessibility landscape of chronic lymphocytic leukemia.*
**a)** Study outline illustrating the use of ATAC-seq for establishing patient-specific and cohort-level maps of chromatin accessibility in CLL
**b)** Saturation analysis showing the number of unique regions of accessible chromatin detected across 88 samples and with a total sequencing depth of 2.25 billion ATAC-seq fragments. Blue and green corridors indicate 95% confidence intervals for samples added in random order over 1,000 iterations.
**c)** Representative genome browser plot showing ATAC-seq signal in individual CLL samples (top), average signal across the cohort (center), and reference data from the ENCODE project (bottom). Browsable tracks are available from the supplementary website (<http://rendeiro2016.computational-epigenetics.org/>).
**d)** Absolute (frequency) and relative (enrichment) co-localization of CLL open chromatin regions with gene annotations (left) and chromatin state segmentations for CD19+ B cells according to the Roadmap Epigenomics project (right).

### Figure: {#Figure2}
![](figures/Figure2.pdf){width=70%}
Caption: *Heterogeneity in the CLL chromatin accessibility landscape.*
**a)** Histogram showing the distribution of variance in chromatin accessibility for promoter regions and putative distal regulatory regions across all genes and for a set of 81 CLL-relevant genes. Chromatin accessibility scores were averaged across all regulatory regions assigned to a given gene.
**b)** Violin plots of chromatin accessibility at the promoters and enhancers of genes with a known role in B cell biology and/or CLL pathogenesis.
**c)** ATAC-seq signal across the CLL cohort in the vicinity of two selected genes.
**d)** Scatterplot of the change in variance-to-mean ratio in chromatin-accessible regions between the mCLL and uCLL sample groups (x-axis) and the significance of variability of each region within each group (y-axis).
**e)** Scatterplots of mean accessibility (left) and variance-to-mean ratio within each sample group (right) with yellow dots indicating significantly variable regions within each sample group. The plot on the left illustrates how significantly variable regions are dispersed across the accessibility range and are thus not strongly coupled with differences in mean accessibility between the groups.
**f)** Most highly enriched region sets that significantly overlap the variable regions within the mCLL (blue) and uCLL (orange) sample groups, based on LOLA analysis[@Sheffield2015].

### Figure: {#Figure3}
![](figures/Figure3.pdf){width=60%}
Caption: *Disease subtype-specific patterns of chromatin accessibility.*
**a)** Methodology for deriving disease subtype-specific patterns of chromatin accessibility: A machine learning algorithm is trained to distinguish between different sample groups (e.g., IGHV mutated vs. unmutated), and its performance is evaluated by cross-validation. The most predictive features are identified by feature extraction from aggregation of the cross-validated models and clustered into differential chromatin signatures.
**b)** ROC curve summarizing the prediction performance of a random forest classifier that accurately distinguishes between IGHV unmutated and IGHV mutated samples. The ROC curve was produced by leave-one-out cross-validation using the full dataset of 112,168 accessible chromatin regions as input features. “AUC” refers to the ROC area under curve measure of prediction performance, and sensitivity/specificity values are shown for the point on the ROC curve that is closest to the top left corner. The grey line indicates the expected performance of a random classifier, corresponding to an AUC of 0.5).
**c)** Clustered heatmap for the most discriminating regions between IGHV mutated (mCLL) and unmutated (uCLL) samples.
**d)** Co-localization of mCLL (blue) and uCLL (green) associated regions with gene annotations (left) and chromatin state segmentations for CD19+ B cells obtained from the Roadmap Epigenomics project (right).
**e)** Most enriched region sets for the mCLL (blue) and uCLL (green) clusters, based on LOLA analysis[@Sheffield2015].
**f)** Most enriched cellular pathways among the mCLL (blue) and uCLL (green) linked genes, based on Reactome and KEGG databases.
**g)** Hierarchical clustering of all CLL samples based on the sample-wise correlation of chromatin accessibility for the most discriminatory regions defined in **c)**. Samples are colored according to IGHV mutation status (top) and cluster assignment (bottom).
**h)** Principal component analysis of all CLL samples based on the same regions and using the same color code as in **g)**.

### Figure: {#Figure4}
![](figures/Figure4.pdf){width=100%}
Caption: *Gene regulatory network inference from transcription factor footprints in chromatin accessibility data.*
**a)** Methodology for inferring gene regulatory networks from ATAC-seq data using transcription factor (TF) footprinting, mapping of transcription factor to genes, and network reconstruction.
**b)** Most differentially regulated genes between the networks inferred for uCLL and mCLL, respectively. Node size reflects the node’s total number of connections, and color indicates the network in which the node has more connections (uCLL: orange; mCLL: blue).
**c)** Relative change in the number of connections between the mCLL and uCLL networks for all genes.
**d)** Relative change in the number of connections between the mCLL and uCLL networks, focusing on a selection of 81 genes with a known role in B cell biology and/or CLL pathogenesis.

## Supplemental information

#### Supplementary Table 1
Sample-wise clinically relevant annotation for all samples in this study.

#### Supplementary Table 2
Summary statistics of the ATAC-seq experiments. Descriptive statistics on the total number of sequenced, duplicate, and filtered reads for each sample in this study. 2.25 billion ATAC-seq fragments were sequenced in total, which corresponds to an average sequencing depth of 25.3 million sequenced fragments per sample. The large percentage of reads mapping to the mitochondrial genome is a well-known characteristic of ATAC-seq (these reads were excluded from the analysis).

#### Supplementary Figure 1
*The cohort is broadly representative of CLLs encountered in clinical care.*

The diagram summarizes the clinical data for the 88 samples included in the ATAC-seq cohort.

#### Supplementary Figure 2
*ATAC-seq fragment length distribution indicates high data quality.*

Distribution of ATAC-seq fragment lengths observed for published GM12878 data[@Buenrostro2013] and for four randomly selected CLL samples from this study. Fragment lengths were inferred based on paired-end sequencing data. The characteristic patterns of nucleosome-associated fragment lengths are observed in all cases.

#### Supplementary Figure 3
*For each sample the ATAC-seq libraries identify most open chromatin regions.*

Relationship of the number of sequenced reads and the number of detected open chromatin regions. Green corridors indicate 95% confidence intervals for random subsampling across all samples; the blue line shows the mean.

#### Supplementary Figure 4
*CLL chromatin accessible regions are enriched for enhancers and promoters.*

**a)** Histogram showing the number of samples in which a given open chromatin region from the CLL consensus map has been detected as a significant ATAC-seq peak.
**b)** Frequency and enrichment for genomic region types in the CLL consensus map compared to region sets of identical size and lengths that were randomized 1,000 times across the genome.
**c)** Frequency and enrichment for CD19+ chromatin states calculated as in panel **b)**.

#### Supplementary Figure 5
*Heterogeneity in chromatin accessibility affects B cell and CLL related genes.*

Violin plots of normalized chromatin accessibility values for gene promoters (regions located within 2 kilobases of the transcription start site) and distal regulatory elements (regions located at least 2 kilobases away from the nearest transcription start site) for 81 genes with a known role in B cell biology and/or CLL pathogenesis.

#### Supplementary Figure 6
*Genomic regions linked to IGHV mutation status are variable and predictive.*

**a)** Distribution of dynamic range, cohort support, and variance-to-mean ratio for all CLL open chromatin regions and for the IGHV-specific regions that were identified by the classifier.
**b)** Annotated hierarchical clustering of all CLL samples based on the sample-wise correlation of chromatin accessibility at IGHV-specific regions.

#### Supplementary Figure 7
*Cohort-level gene regulatory network identifies regulator proteins in CLL.*

**a)** Gene regulatory network of CLL inferred from footprint predictions of transcription factor binding, based on the ATAC-seq data of all CLL samples. Only nodes with more than 200 connections are displayed.
**b)** Number of connections across all genes in the cohort-level gene regulatory network (transcription factors are shown in red).

#### Supplementary Figure 8
*Gene regulatory networks of IGHV mutated and unmutated are globally similar.*

Gene regulatory networks inferred based on the IGHV unmutated samples (uCLL, left) and based on the IGHV mutated samples (mCLL, right). Only nodes with more than 200 are displayed.

#### Supplementary Figure 9
*Disease subtype-specific gene regulatory networks are locally highly distinct.*

Subnetworks with the network neighbors of PAX9 and CD22, shown separately for IGHV unmutated samples (uCLL, left) and IGHV mutated samples (mCLL, right). Edge widths indicate the strength of the interactions.

## Experimental Procedures
### Sample acquisition and clinical data
All patients were diagnosed and treated at the Royal Bournemouth Hospital (UK) according to the revised guidelines of the International Workshop Chronic Lymphocytic Leukemia/National Cancer Institute (IWCLL/NCI). Patients were selected to reflect the clinical and biological heterogeneity of the disease. Sequential samples were included for some of the patients, in order to assess the development of chromatin accessibility patterns over time. All samples contained more than 80% leukemic cells. Established chromosomal rearrangements were diagnosed by fluorescence in situ hybridization (Abbott Diagnostics; DakoCytomation) or multiple ligation dependent probe amplification using the MLPA P037 CLL-1 probemix (MRC Holland SALSA) according to the manufacturers’ instructions. Chromosomal analysis was performed and described according to the International System for Human Cytogenetic Nomenclature. ZAP70 and CD38 expression was determined as previously described, where 10% and 30% positive cells were classed as positive, respectively[@Best2006; @Oscier2002]. IGHV genes were sequenced as previously described[@Hamblin1999], and a threshold of >98% germline homology was taken to define the unmutated subset[@Hamblin1999]. The study was approved and overseen by the local ethics committee of the contributing institutions.

### Chromatin accessibility mapping by ATAC-seq
Accessible chromatin mapping was performed using the ATAC-seq method as described[@Buenrostro2013] with minor adaptations. In each experiment, 10^5^ cells were washed once in 50 $mu$l PBS, resuspended in 50 $mu$l ATAC-seq lysis buffer (10 mM Tris-HCl, pH 7.4, 10 mM NaCl, 3 mM MgCl2, and 0.1% IGEPAL CA-630) and centrifuged for 10 min at 4°C. Upon centrifugation, the pellet was washed briefly in 50 $mu$l MgCl2 buffer (10 mM Tris, pH 8.0, and 5 mM MgCl2) before incubating in the transposase reaction mix (12.5 $mu$l 2x TD buffer, 2 $mu$l transposase (Illumina) and 10.5 $mu$l nuclease-free water) for 30 min at 37°C. After DNA purification with the MinElute kit, 1 $mu$l of the eluted DNA was used in a qPCR reaction to estimate the optimum number of amplification cycles. Library amplification was followed by SPRI size selection to exclude fragments larger than 1,200 bp. DNA concentration was measured with a Qubit fluorometer (Life Technologies). Library amplification was performed using custom Nextera primers as described previously22, and the libraries were sequenced by the Biomedical Sequencing Facility at CeMM using the Illumina HiSeq3000/4000 platform and the 25bp paired-end configuration. 

### Preprocessing of the ATAC-seq data
Reads were trimmed using Skewer[@Jiang2014]. Trimmed reads were aligned to the GRCh37/hg19 assembly of the human genome using Bowtie2[@Langmead2012] with the “–very-sensitive” parameter and the aligner’s default. Duplicate reads were removed using sambamba markdup, and only properly paired reads with mapping quality >30 and alignment to the nuclear genome were kept. All downstream analyses were performed on the filtered reads. Genome browser tracks were created with the genomeCoverageBed command in BEDTools[@Quinlan2010] and normalized such that each value represents the read count per base pair per thousand filtered reads. Finally, the UCSC Genome Browser’s bedGraphToBigWig tool was used to produce a bigWig file. The tracks with percentile signal across the cohort were created by quantifying the ATAC-seq read coverage at every reference genome position using BEDTools coverage and normalizing it between samples. The normalization was done by dividing each value by the total number of filtered reads and multiplying by 10 million, in order to obtain numbers that are comparable and easy to visualize. Next, the mean as well as the 5th, 25th, 75th and 95th percentiles of signal across the whole cohort were calculated with Numpy, converted into bedgraph files, and subsequently to bigwig format using UCSC’s bedGraphToBigWig. Peak calling was performed with MACS2[@Zhang2008] using the “–nomodel” and “–extsize 147” flags and arguments, and peaks overlapping blacklisted features as defined by the ENCODE project[@Hoffman2013] were discarded.

### Bioinformatic analysis of chromatin accessibility
The CLL consensus map was created by merging the ATAC-seq peaks from all samples using the BEDTools[@Quinlan2010] *merge* command. To produce [#Figure1]b, we counted the number of unique chromatin-accessible regions after merging peaks for each sample in an iterative fashion, randomizing the sample order 1,000 times and computing 95% confidence intervals across all iterations. The chromatin accessibility of each region in each sample was quantified using Pysam, counting the number of reads from the filtered BAM file that overlapped each region. To normalize read counts across samples, we performed quantile normalization using the *normalize.quantiles* function from the preprocessCore package in R. For each genomic region we calculated the support as the percentage of samples with a called peak in the region, and four measures of ATAC-seq signal variation across the cohort: mean signal, standard deviation, variance-to-mean ratio, and the squared coefficient of variation (the square of the standard deviation over the mean). Additionally, we used BEDTools intersect to annotate each region with the identity of and distance to the nearest transcription start site and the overlap with Ensembl gene annotations (promoters were defined as the 2.5 kb region upstream of the transcription start site). The annotation of chromatin states was based on the 15 state genome segmentation for CD19+ B cells from the Roadmap Epigenomics Project[@Ernst2015] (identifier: E032). To summarize the chromatin accessibility signals into one value per gene (Figure [#Figure2]a, [#Figure2]b and Supplementary Figure 5) we used the accessibility values of the closest region within 1 kb to represent promoter accessibility, and the mean values of all distal regions (more than 2 kb away from the transcription start site) of each gene to represent distal regulatory elements. To investigate variability within the mCLL and uCLL sample groups, we divided the samples in two groups based on their IGHV mutation status (samples below a 98% homology threshold were considered mutated) and used the F test from the var.test function in R on the chromatin accessibility values of all CLL cohort regions. Significantly variable regions were defined as having a Bonferroni-corrected p-value below 0.05 and mean accessibility above 1. Region set enrichment analysis was performed on the significantly variable regions of each group using LOLA[@Sheffield2015] and its core databases: transcription factor binding sites from ENCODE[@Hoffman2013], tissue clustered DNase hypersensitive sites[@Sheffield2013], the Codex database[@Sanchez-Castillo2015], UCSC Genome Browser annotation tracks[@Rosenbloom2015], the Cistrome database[@Liu2011], and data from the BLUEPRINT project[@Adams2012].

### Machine learning analysis of disease subgroups
A random forest classifier from the scikit-learn[@Pedregosa2011] Python library (*sklearn.ensemble.RandomForestClassifier*) was trained with the samples’ IGHV mutation status as class label and the chromatin accessibility values for each sample at each of the 112,168 consensus regions as input features (prediction attributes). All samples with known IGHV mutation status were used for class prediction, the performance was evaluated by leave-one-out cross-validation, and the results were plotted as ROC curves using scikit-learn. The most predictive regions for IGHV mutation status were selected by averaging the feature importance of the random forest classifier over all iterations and selecting all features with non-zero importance. Region set enrichment was performed using LOLA[@Sheffield2015] as described above. Pathway enrichment analysis was performed in R using seq2pathway[@Wang2015] . The sample clustering in Figure [#Figure3]g was based on the pairwise correlation of ATAC-seq signal in the predictive regions between samples, and the dendrogram was plotted using Scipy’s hierarchical clustering function. With the same values of chromatin accessibility from above, we performed principal component analysis on the CLL samples using R’s implementation in the *prcomp* function.

### Gene regulatory network inference from TF footprints
Transcription factor binding maps as the basis for inferring gene regulatory networks were derived by footprinting analysis using the PIQ software[@Sherwood2014] and a set of 366 Human transcription factor motifs from the JASPAR database. As described previously[@Qu2015], we retained only those transcription factors with at least 500 high-purity (>0.7) binding sites overlapping with an ATAC-seq peak. Our scheme for assigning transcription factor binding sites to genes was as follows: Binding sites located in the gene body or in the 2.5 kb region upstream of its transcription start site were assigned to the overlapping gene(s). Intergenic binding sites were assigned to the gene whose transcription start site was closest to the peak. For this association, we used the Ensembl gene annotation version 75, and we treated non-protein-coding genes in the same way as protein-coding genes. To infer gene regulatory networks, an interaction score was calculated in a similar way as described previously[@Qu2015]: The interaction score between a transcription factor $t$ and a gene $g$ ($S_{t,g}$)  was defined as the sum over all $n$ transcription factor binding sites of $t$ that can be assigned to $g$: 

~~~math #interaction-score
S_{t,g} = \sum_{i=0}^{n}2 \times (P_{i} - 0.5) * 10 ^{-(\frac{d_{i, g}}{100000})}
~~~

In this formula $P_{i}$ is the PIQ purity score and $d_{i, g}$ is the distance of a particular transcription factor binding site $i$ to gene $g$. This score establishes a unidirectional (transcription factors to genes) and weighted (based on the interaction score) relationship, which form the edges of the gene regulatory network. We inferred gene regulatory networks for all samples combined and also for the two disease subtypes (uCLL and mCLL) based on IGHV mutation status. We considered only transcription factor to gene interactions with scores above 1, and in Figure [#Figure4]b as well as Supplementary Figures 7 and 8 we plotted only nodes with more than 200 connections. In order to compare the inferred IGHV-unmutated and IGHV-mutated networks, we divided the degree of each node by the total number of edges in each network to compensate for differences in the absolute number of detected interactions, and we quantified differences by subtracting and log2-transforming this value between networks for each node.


### Data availability
The ATAC-seq intensity data are available for interactive browsing and download from the supplementary website (<http://rendeiro2016.computational-epigenetics.org/>). They are also being submitted to NCBI GEO (accession number pending). The raw sequencing data are being submitted to the EBI EGA for controlled access, in order to protect the privacy of the patients who have donated the samples.

## References
