<!DOCTYPE html>
<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
<!--[if IE 7]>         <html class="no-js lt-ie9 lt-ie8"> <![endif]-->
<!--[if IE 8]>         <html class="no-js lt-ie9"> <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js"> <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <meta name="generator" content="scholpandoc">
  <meta name="viewport" content="width=device-width">
  
  <meta name="author" content="André Rendeiro">
  <meta name="dcterms.date" content="2015-08-02">
  <title>CLL literature review</title>
  <style type="text/css">code{white-space: pre;}</style>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.7.1/modernizr.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/respond.js/1.4.2/respond.js"></script>
  <link rel="stylesheet" href="style.css">
</head>
<body>
<div class="scholmd-container">
<div class="scholmd-main">
<div class="scholmd-content">
<header>
<h1 class="scholmd-title">CLL literature review</h1>
<div class="scholmd-author">
André Rendeiro
</div>
<div class="scholmd-date">August 02, 2015</div>
</header>
<h2 id="what-is-this">What is this:</h2>
<p>This is a short overview of fundamental literature on chronic lymphocytic leukaemia (CLL). While this meant both as a review for me and a <em>get-up-to-speed-quickly</em> work for others, this should not encourage you not to read the actual papers.</p>
<h1 id="basic-review">Basic review</h1>
<h2 id="b-cell-biology">B-cell biology</h2>
<ol type="1">
<li>In secondary lymphoid organs, mature peripheral B cells (follicular (Fo) B cells) acquire antigen from follicular dendritic cells (FDCs) and in turn presents it to cognate CD4+ TfH cells. During TfH-B cell conjugation, the B cell receives critical signals to undergo class switching and to commence monoclonal expansion.</li>
<li>After several rounds of cellular division, the B cells go through somatic hypermutation, generating a diversity of clones in the germinal center.</li>
<li>Upon receiving an unidentified stimulus, the maturing B cells migrate from the dark zone to the light zone and start to express their antibody on the cell surface. They are in a state of activated apoptosis and compete for survival signals from follicular dendritic cell and TfH cells. This rescue process is dependent on the antibody affinity to the antigen. T cells are also believed to prevent the generation of autoreactive germinal center B cells.</li>
<li>Maturing B cells receive a final differentiation signal to exit the germinal center as plasma cells or memory B cells.</li>
</ol>
<h2 id="why-is-there-cll">Why is there CLL?</h2>
<p>Mistargeted somatic hypermutation is a likely mechanism in the development of CLL and other B-cell lymphomas.</p>
<h2 id="clinical-staging">Clinical staging</h2>
<p>Based primarily on the presence of a low platelet or red cell count.</p>
<h4 id="rai-staging-system">Rai Staging System</h4>
<p>Absolute values of lymphocytosis (&gt;15,000/mm3) are considered in all stages.</p>
<ul>
<li>Stage 0: no adenopathy, hepatosplenomegaly, anemia, or thrombocytopenia.</li>
<li>Stage I: lymphadenopathy without hepatosplenomegaly, anemia, or thrombocytopenia.</li>
<li>Stage II: either hepatomegaly or splenomegaly with or without lymphadenopathy.</li>
<li>Stage III: anemia (hemoglobin &lt;11 g/dL) with or without lymphadenopathy, hepatomegaly, or splenomegaly.</li>
<li>Stage IV: characterized by absolute lymphocytosis and thrombocytopenia (&lt;100,000/mm3) with or without lymphadenopathy, hepatomegaly, splenomegaly, or anemia.</li>
</ul>
<h4 id="binet-classification">Binet Classification</h4>
<ul>
<li>Clinical stage A: no anemia or thrombocytopenia and fewer than three areas of lymphoid involvement (Rai stages 0, I, and II).</li>
<li>Clinical stage B: no anemia or thrombocytopenia with three or more areas of lymphoid involvement (Rai stages I and II).</li>
<li>Clinical stage C: anemia and/or thrombocytopenia regardless of the number of areas of lymphoid enlargement (Rai stages III and IV).</li>
</ul>
<h2 id="molecular-classification">Molecular classification</h2>
<p>Compiled from various publications (see )</p>
<ul>
<li>CD5+ B-cells</li>
<li>IGHV unmutated CLL (u-CLL) more like naïve B-cells
<ul>
<li>Likely comes from pre-germinal center B cells</li>
<li>Enriched in high CD38 and ZAP70 (T cell receptor) expression</li>
<li>Enriched in NOTCH1 mutations</li>
<li>Poor prognosis</li>
</ul></li>
<li>IGHV mutated CLL (m-CLL) more like memory B-cells
<ul>
<li>Likely comes from post-germinal center, mature B cells</li>
<li>More patients left untreated</li>
</ul></li>
<li>there are obvoulsly intermediate states as well (i-CLL)</li>
</ul>
<blockquote>
<p><strong>personal note</strong> although the intermediate group has obviously features of both, it seems more similar to the memory B cell (m-CLL) group</p>
</blockquote>
<h2 id="classical-markers">Classical markers</h2>
<ul>
<li>IGHV mutational status (&lt;98% homology is considered mutated)</li>
<li>cytogenetics</li>
<li>CD38 expression</li>
<li>ZAP-70 expression (predictive of IGHV mutation status)</li>
</ul>
<p>These are nowadays considered <strong>not fundamental</strong> for the initiation or continuation of leukemic process, but are still <strong>descriptive</strong> of it.</p>
<h2 id="b-cell-surface-markers">B-cell surface markers</h2>
<ul>
<li>All B-cells: CD19+</li>
<li>naive B cells (NBCs): CD19+, CD27−, IgD+</li>
<li>CD5+ naive B cells: CD19+, CD5+, CD27–, CD38low</li>
<li>non-class-switched memory B cells (ncsMBCs): CD19+, CD27+, IgM+ or IgD+ (both to sort)</li>
<li>class-switched memory B cells (csMBCs): CD19+, CD27+, IgA+ or IgG+</li>
</ul>
<h1 id="recent-papers">Recent Papers</h1>
<h2 id="puente-et-al-nature-2011"><a href="http://doi.org/doi:10.1038/nature10113">Puente et al, Nature, 2011</a></h2>
<blockquote>
<p>Puente, X. S., Pinyol, M., Quesada, V., Conde, L., Ordóñez, G. R., Villamor, N., … Campo, E. (2011). <strong>Whole-genome sequencing identifies recurrent mutations in chronic lymphocytic leukaemia</strong>. Nature, 475(7354), 101–105. doi:10.1038/nature10113</p>
</blockquote>
<h3 id="design">Design</h3>
<ul>
<li>WGS of 4 CLLs:
<ul>
<li>2 IGVH mutated/ 2 unmutated</li>
<li>no TP53 mutations</li>
<li>Patients were already in advanced stages - no treatments so far. Choice was made to avoid cases with obvious mutations and focus on discovery, so these 4 cases are probably not be the most representative of the disease.</li>
</ul></li>
<li>Validation cohort:
<ul>
<li>363 patients</li>
<li>samples (&gt; 70% CLL cells) + matched controls (&lt; 5% CLL cells)</li>
</ul></li>
</ul>
<h3 id="results">Results</h3>
<ul>
<li>~1000 mutations per patient (&lt; 1/Mb)</li>
<li>A&gt;C/T&gt;G mutations are more frequent in CLL-M. This is likely due to polymerase η (Eta) action in immunoglobulin genes. Pol η causes mutations that occur in clusters with bias to certain bases (generally more A&gt;C/T&gt;G).</li>
</ul>
<blockquote>
<p>Side note: somatic hypermutation is not rare and genes such as BCL6, MYC and PIM1, are recurrently mutated by somatic hypermutation in different lymphomas <a href="http://doi.org/doi:10.1038/35085588">Pasqualucci et al, 2001, Nature</a>.</p>
</blockquote>
<h4 id="recurrently-mutated-genes-with-direct-functional-implications-for-the-disease">4 recurrently mutated genes with direct functional implications for the disease:</h4>
<ul>
<li>NOTCH1
<ul>
<li>frequency: 12%</li>
<li>Early stop codon mutations</li>
<li>activating mutation</li>
<li>Notch pathway significantly affected (upregulated)</li>
<li>Clinicaly advanced, less survival</li>
</ul></li>
<li>MYD88
<ul>
<li>frequency: 2.9%</li>
<li>MYD88 mutations are relevant in other lymphoid neoplasias</li>
<li>activating mutation</li>
<li>higher MYD88 phosphorilation, higher NFKb binding</li>
<li>stimulation of IL-1r or TLR receptors in MYD88-mutated cases yielded 5-150 fold secretion of IL-1r antagonist, IL-6, chemokine ligands when compared to unmutated MYD88. same stimulation in cells mutated in MYD88 with inactivating mutation resulted in no response. Production of these cytokines has been implicated in recruitment of T cells and macrophages. Recruitment of these cells might be creating a favourable environment for CLL survival</li>
<li>MYD88-mutated patients had earlier diagnose data and more advanced disease</li>
<li>almost all MYD88-mutated patients were IGHV-mutated</li>
</ul></li>
<li>XPO1
<ul>
<li>frequency: 2.4%</li>
<li>exportin 1 (nuclear export)</li>
<li>IGHV-unmutated exclusive</li>
<li>co-occurent with NOTCH1 mutations</li>
</ul></li>
<li>KLHL6
<ul>
<li>frequency: 1.8%</li>
<li>implicated in the formation of germinal centres (B-cell maturation site)</li>
<li>almost exclusive to IGHV-mutated (mutations have pol eta pattern)</li>
</ul></li>
</ul>
<h4 id="other-mutationsaberations">Other mutations/aberations</h4>
<ul>
<li>del 13q14</li>
<li>del 6q14-q22</li>
<li>del CCND2 (cyclin D2)</li>
</ul>
<h2 id="braggio-et-al-leukemia-2012"><a href="http://doi.org/doi:10.1038/leu.2012.14">Braggio et al, Leukemia, 2012</a></h2>
<blockquote>
<p>Braggio, E., Kay, N. E., VanWier, S., Tschumper, R. C., Smoley, S., Eckel-Passow, J. E., … Fonseca, R. (2012). <strong>Longitudinal genome-wide analysis of patients with chronic lymphocytic leukemia reveals complex evolution of clonal architecture at disease progression and at the time of relapse</strong>. Leukemia, 26(7), 1698–1701. doi:10.1038/leu.2012.14</p>
</blockquote>
<p>This paper is about structural aberrations during CLL clonal evolution.</p>
<h3 id="design-1">Design</h3>
<ul>
<li>22 patients, at least two timepoints</li>
<li>aCGH</li>
<li>samples grouped acording to time-points:
<ul>
<li>TP1: collected &gt;6 months before starting first-line treatment;</li>
<li>TP2: collected at the time of progression, during treatment;</li>
<li>TP3: collected &gt;6 months after initial treatment but before subsequent treatments</li>
</ul></li>
</ul>
<h3 id="results-1">Results</h3>
<ul>
<li>Only 6 of 22 patients showed copy-number abnormalities differences between time-points.</li>
<li>These were validated with FISH</li>
<li>Trisomy 12 and deletion 11q32 were stable over time</li>
<li>Increment in the number of cases with deletion 13q14.3 and 17p</li>
<li>Four patients had subclones. In all cases, the dominant clone at TP1 or TP2 was no longer at TP3:
<ul>
<li>One patient had 3 subclones, all sharing common aberations like trisomy 12, showing that they occurred probably after the emergence of a clone. Clone “A” dominated in TP1, was only ~70% of cells in TP2 and 20% at TP3. At TP3 subclone “B” was 60% of all cells and subclone “C” 20%.</li>
<li>Another patient had two subclones, one of which had 70% dominance at TP2, but only 10% at TP3.</li>
</ul></li>
</ul>
<p>CLL progression can occur in either a linear or a branching manner, with multiple genetic subclones evolving either in succession or in parallel</p>
<h2 id="quesada-et-al-nature-genetics-2012"><a href="http://doi.org/doi:10.1038/ng.1032">Quesada et al, Nature Genetics, 2012</a></h2>
<blockquote>
<p>Quesada, V., Conde, L., Villamor, N., Ordóñez, G. R., Jares, P., Bassaganyas, L., … López-Otín, C. (2011). Exome sequencing identifies recurrent mutations of the splicing factor SF3B1 gene in chronic lymphocytic leukemia. Nature Genetics, 44(1), 47–52. doi:10.1038/ng.1032</p>
</blockquote>
<h3 id="design-2">Design</h3>
<ul>
<li>105 exomes matched tumour-normal:
<ul>
<li>60 mCLL</li>
<li>45 uCLL ### Results</li>
</ul></li>
</ul>
<h2 id="doménech-et-al-plos-one-2012"><a href="http://doi.org/doi:10.1371/journal.pone.0038158">Doménech et al, PLoS ONE, 2012</a></h2>
<blockquote>
<p>Doménech, E., Gómez-López, G., Gzlez-Peña, D., López, M., Herreros, B., Menezes, J., … Sánchez-Beato, M. (2012). New mutations in chronic lymphocytic leukemia identified by target enrichment and deep sequencing. PLoS ONE, 7(6), 2–7. doi:10.1371/journal.pone.0038158</p>
</blockquote>
<h3 id="design-3">Design</h3>
<h3 id="results-2">Results</h3>
<h2 id="kulis-et-al-nature-genetics-2012"><a href="http://doi.org/doi:10.1038/ng.2443">Kulis et al, Nature Genetics, 2012</a></h2>
<blockquote>
<p>Kulis, M., Heath, S., Bibikova, M., Queirós, A. C., Navarro, A., Clot, G., … Martín-Subero, J. I. (2012). <strong>Epigenomic analysis detects widespread gene-body DNA hypomethylation in chronic lymphocytic leukemia</strong>. Nature Genetics, 44(11), 1236–1242. doi:10.1038/ng.2443</p>
</blockquote>
<h3 id="design-4">Design</h3>
<ul>
<li>WGBS of two patients (one u-CLL, one m-CLL)</li>
<li>naïve B-cells, non-class-switched memory B-cells (ncsMBC) and class-switched memory B-cells (csMBC) + low coverage WGS from a single donor</li>
<li>139 CLLs in 450k arrays + expression arrays</li>
<li>450k array on various B cell populations from age matched healty donors
<ul>
<li>19 whole B-cell (CD19)</li>
<li>three each of CD5+ NBC, NBC, ncsMBCs and csMBCs with &gt; 95% purity</li>
</ul></li>
<li>exome sequecing of 88 CLLs from previous study</li>
</ul>
<h3 id="results-3">Results</h3>
<ul>
<li>NBCs vs csMBCs: 1,076,208 DMRs in 19125 genes</li>
<li>96% of DMRs were hypomethylated</li>
<li>little difference between ncsMBCs and csMBCs</li>
<li>most DMRs in gene bodies and intergenic space (outside CGIs)</li>
<li>PCA distinguishes u-CLL and m-CLL</li>
<li>poor overal correlation between DNAme and expression</li>
<li>~3,265 DMRs between u-CLL and m-CLL detected in the arrays, &gt; 2.5 million in WGBS</li>
<li>~50% of the DMRs were not distinguishable between uCLL and NBCs or mCLL and scMBCs</li>
<li>differences between uCLL and mCLL in array data:
<ul>
<li>3,243 hypermethylated and 29,743 hypomethylated CpGs in U-CLL compared with NBCs and CD5+ NBCs</li>
<li>246 hypermethylated and 4,606 hypomethylated CpGs in M-CLL compared with MBCs</li>
</ul></li>
<li>higher differences in the WGBS set:
<ul>
<li>1,838,346 DMRs between U-CLL and NBCs (1,779,168 hypomethylated and 59,178 hypermethylated)</li>
<li>1,254,527 DMRs between M-CLL and csMBCs (982,683 hypomethylated and 271,844 hypermethylated)</li>
</ul></li>
<li>in general hypermethylation was at 5’ regions of genes and CGIs, whereas hypomethylation was mainly at gene bodies</li>
<li>hypomethylated regions are enriched in signaling processes such as B-cell receptor–, NF-κB– and calcium-activated pathways; T-cell costimulation; and cytokine-cytokine receptor interactions</li>
<li>CpGs with significant correlation with expression (4%) were enriched in gene bodies</li>
<li>in a fraction, correlation between gene expression and DNA methylation was simultaneously negative in 5′ regions and positive in gene bodies</li>
<li>CpGs hypomethylated in U-CLL or M-CLL were enriched for enhancer regions in B cells and IRF and OCT1 binding sites</li>
<li>enhancers were also enriched in the uCLL/NBC comparison</li>
<li>gene body CpGs with negative correlation with expression were enriched in enhancers, whereas positively correlated ones had only slight enrichment<br /></li>
<li>tried to loook at the relationship between DNAme and splicing aberations - only two genes showed DNA methylation correlated with exon skipping, and lack of methylation correlated with exon inclusion</li>
<li>Eight recurrently mutated genes in CLL showed aberrant hypomethylation (ASXL1, BRAF, CDH23, EGR2, FAM117A, PHC2, POT1 and SF3B1) - not more than expected by chance - exome data only used here</li>
<li>Further clustering of samples based on NBC- and MBC-CpGs showed three clusters, the third one with a majority of mCLLs enriched in fewer IGHV mutations. These CLLs could come from “antigen- experienced, germinal center–independent B cell with low levels of somatic hypermutation”.</li>
<li>The three groups showed significant clinical features between them</li>
</ul>
<h2 id="queirós-et-al-leukemia-2014"><a href="http://doi.org/doi:10.1038/leu.2014.252">Queirós et al, Leukemia, 2014</a></h2>
<blockquote>
<p>Queirós, a C., Villamor, N., Clot, G., Martinez-Trillos, A., Kulis, M., Navarro, A., … Martín-Subero, J. I. (2014). <strong>A B-cell epigenetic signature defines three biological subgroups of chronic lymphocytic leukemia with clinical impact</strong>. Leukemia, (August 2014), 598–605. doi:10.1038/leu.2014.252</p>
</blockquote>
<h2 id="design-5">### Design</h2>
<h2 id="results-4">### Results</h2>
<h2 id="baliakas-et-al-leukemia-2015"><a href="http://doi.org/doi:10.1038/leu.2014.196">Baliakas et al, Leukemia, 2015</a></h2>
<p>Baliakas, P., Hadzidimitriou, A., Sutton, L., Rossi, D., Minga, E., Villamor, N., … Rosenquist, R. (2014). <strong>Recurrent mutations refine prognosis in chronic lymphocytic leukemia</strong>. Leukemia, (April), 1–8. doi:10.1038/leu.2014.196</p>
<h2 id="design-6">### Design</h2>
<h2 id="results-5">### Results</h2>
<h2 id="puente-et-al-nature-2015"><a href="http://doi.org/doi:10.1038/nature14666">Puente et al, Nature, 2015</a></h2>
<blockquote>
<p>Puente, X. S., Beà, S., Valdés-Mas, R., Villamor, N., Gutiérrez-Abril, J., Martín-Subero, J. I., … Campo, E. (2015). <strong>Non-coding recurrent mutations in chronic lymphocytic leukaemia</strong>. Nature. doi:10.1038/nature14666</p>
</blockquote>
<h2 id="design-7">### Design</h2>
<h2 id="results-6">### Results</h2>
<hr />
<p>Made with <a href="http://scholarlymarkdown.com/">Scholarly Markdown</a>, and <a href="http://scholdoc.scholarlymarkdown.com/">Scholdoc</a></p>
</div>
</div>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
      processClass: "math"
    },
    TeX: {
        TagSide: "left",
        TagIndent: "1.2em",
        equationNumbers: {
            autoNumber: "AMS"
        },
        Macros: {
            ensuremath: ["#1",1],
            textsf: ["\\mathsf{\\text{#1}}",1],
            texttt: ["\\mathtt{\\text{#1}}",1]
        }
    },
    "HTML-CSS": { 
        scale: 100,
        availableFonts: ["TeX"], 
        preferredFont: "TeX",
        webFont: "TeX",
        imageFont: "TeX",
        EqnChunk: 1000
    }
});
</script>
</div>
</body>
</html>
